// @byted-icube/devtools-i18n/dist/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.initI18n = exports.i18n = exports.I18nNamespace = void 0;
const devtools_base_1 = require("@byted-icube/devtools-base");
const resource_1 = require("./resource");
var I18nNamespace;
(function (I18nNamespace) {
    I18nNamespace["AI"] = "ai";
    I18nNamespace["DevTools"] = "devtools";
})(I18nNamespace || (exports.I18nNamespace = I18nNamespace = {}));
exports.i18n = (0, devtools_base_1.createI18nTranslator)({
    resources: resource_1.DevToolsResources,
});
async function initI18n(options) {
    const { logger, namespace = I18nNamespace.DevTools, resources } = options;
    let i18nInstance = exports.i18n;
    if (resources) {
        i18nInstance = (0, devtools_base_1.createI18nTranslator)({ resources });
    }
    i18nInstance.setLang(options.lang);
    (0, devtools_base_1.fetchLocaleDatas)({ i18n: i18nInstance, namespace, ...options }).then(() => {
        logger.info(`init i18n success`);
    }).catch((error) => {
        logger.info(`init i18n failed: ${error.stack || error.message}`);
    });
    return i18nInstance;
}
exports.initI18n = initI18n;
__exportStar(require("./resource"), exports);
