// @ai-local-server/common/dist/typings/index.js
import { Region } from '@byted-icube/devtools-base';
export var Product;
(function (Product) {
    Product["MarsCode"] = "marscode";
    Product["CozePlugin"] = "coze-plugin";
    Product["CloudIDE"] = "cloudide";
    Product["Practice"] = "practice";
    Product["Merlin"] = "merlin";
    Product["MarsCodeInternal"] = "marscode-internal";
    Product["DesktopInternal"] = "desktop-internal";
    Product["Course"] = "course";
    Product["AITerminal"] = "ai-terminal";
})(Product || (Product = {}));
export var Language;
(function (Language) {
    Language["ZH"] = "zh";
    Language["EN"] = "en";
})(Language || (Language = {}));
export var Namespace;
(function (Namespace) {
    Namespace["AI"] = "ai";
    Namespace["DevTools"] = "devtools";
    Namespace["ComplianceExtension"] = "compliance-extension";
    Namespace["Coze"] = "coze";
    Namespace["Doc"] = "doc";
    Namespace["iCube"] = "icube";
    Namespace["Stable"] = "stable";
    Namespace["Website"] = "website";
})(Namespace || (Namespace = {}));
export { Region };
export * from './config';
export * from './slardar';
// product 和 region 的组合，跟 local 配置文件一一对应
export var ProductAndRegion;
(function (ProductAndRegion) {
    ProductAndRegion["CloudIDEBOE"] = "CloudIDEBOE";
    ProductAndRegion["CloudIDEBOEI18N"] = "CloudIDEBOEI18N";
    ProductAndRegion["CloudIDECN"] = "CloudIDECN";
    ProductAndRegion["CloudIDEUS"] = "CloudIDEUS";
    ProductAndRegion["CloudIDESG"] = "CloudIDESG";
    ProductAndRegion["MarsCodeCN"] = "MarsCodeCN";
    ProductAndRegion["MarsCodeSG"] = "MarsCodeSG";
    ProductAndRegion["MarsCodeUS"] = "MarsCodeUS";
    ProductAndRegion["MarsCodeBOE"] = "MarsCodeBOE";
    ProductAndRegion["MarsCodeBOEI18N"] = "MarsCodeBOEI18N";
    ProductAndRegion["MarsCodeInternalCN"] = "MarsCodeInternalCN";
    ProductAndRegion["MarsCodeInternalBOE"] = "MarsCodeInternalBOE";
    ProductAndRegion["DesktopInternalCN"] = "DesktopInternalCN";
    ProductAndRegion["DesktopInternalUS"] = "DesktopInternalUS";
    ProductAndRegion["DesktopInternalSG"] = "DesktopInternalSG";
    ProductAndRegion["CozePluginCN"] = "CozePluginCN";
    ProductAndRegion["CozePluginBOE"] = "CozePluginBOE";
    ProductAndRegion["MerlinBOE"] = "MerlinBOE";
    ProductAndRegion["MerlinCN"] = "MerlinCN";
    ProductAndRegion["PracticeBOE"] = "PracticeBOE";
    ProductAndRegion["PracticeCN"] = "PracticeCN";
    ProductAndRegion["CourseBOE"] = "CourseBOE";
    ProductAndRegion["CourseCN"] = "CourseCN";
})(ProductAndRegion || (ProductAndRegion = {}));
