// @slardar/sdk-web/esm/collector/error.js
import { getDefaultBrowser } from '../utils/defaults';
import { addListenerToWindow } from '../utils/misc';
export var ERROR_SUBJECT_NAME = 'err_0';
export var observeError = function (next, tearDown) {
    var window = getDefaultBrowser();
    if (!window)
        return;
    var removeListener = addListenerToWindow(window, 'error', next, true);
    tearDown(function () {
        removeListener();
    });
};
export var errorSubject = [ERROR_SUBJECT_NAME, observeError];
export var PROMISE_REJECTION_SUBJECT_NAME = 'perr_0';
export var observePromiseRejection = function (next, tearDown) {
    var window = getDefaultBrowser();
    if (!window)
        return;
    var removeListener = addListenerToWindow(window, 'unhandledrejection', next, true);
    tearDown(function () {
        removeListener();
    });
};
export var promiseRejectionSubject = [
    PROMISE_REJECTION_SUBJECT_NAME,
    observePromiseRejection,
];//# sourceMappingURL=https://trae.private/sourcemaps/424b4bd987c6c6a4cadf1a08da420593cc6bf75a/node_modules/@slardar/web/node_modules/@slardar/sdk-web/esm/collector/error.js.map
