// @byted-icube/devtools-base/dist/common/utils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getTeaConfig = exports.getSlardarBid = exports.getDefaultRegion = exports.getEnviromentVariables = void 0;
const region_1 = require("./region");
const tenant_1 = require("./tenant");
const types_1 = require("../types");
const constants_1 = require("../constants");
function getEnviromentVariables() {
    if (typeof process !== 'undefined') {
        return process.env || {};
    }
    else if (typeof window !== 'undefined') {
        return window.env || {};
    }
    return {};
}
exports.getEnviromentVariables = getEnviromentVariables;
function getDefaultRegion() {
    const env = getEnviromentVariables();
    return (0, region_1.getRegion)(env);
}
exports.getDefaultRegion = getDefaultRegion;
function getSlardarBid(options = {}) {
    const env = getEnviromentVariables();
    const region = options.region || getDefaultRegion();
    const tenant = options.tenant || (0, tenant_1.getTenant)(env);
    const inner = tenant === types_1.Tenant.BYTEDANCE;
    return constants_1.DEFAULT_SLARDAR_BID_MAP[region][inner ? 'inner' : 'outer'];
}
exports.getSlardarBid = getSlardarBid;
function getTeaConfig(options = {}) {
    const env = getEnviromentVariables();
    const region = options?.region || getDefaultRegion();
    const tenant = options?.tenant || (0, tenant_1.getTenant)(env);
    const innerOrOuter = tenant === types_1.Tenant.BYTEDANCE ? 'inner' : 'outer';
    return {
        appId: constants_1.TEA_APP_ID_MAP[region][innerOrOuter],
        channel: constants_1.TEA_CHANNEL_MAP[region][innerOrOuter],
        channelDomain: constants_1.TEA_AB_CHANNEL_DOMAIN_MAP[region][innerOrOuter],
        serviceOrigin: constants_1.TEA_AB_SERVICE_ORIGIN_MAP[region][innerOrOuter],
    };
}
exports.getTeaConfig = getTeaConfig;
