// @slardar/sdk-web/esm/monitors/performance/util.js
import { __assign } from "tslib";
/* eslint-disable @typescript-eslint/prefer-optional-chain */
import { noop } from '@slardar/sdk-template';
import { PERFORMANCE_METRIC_EV_TYPE } from '../../constants';
import { getDefaultDocument, addListenerToDocument } from '../../utils';
export var defaultMetricContext = {
    isSupport: true,
    isPolyfill: false,
    isBounced: false,
    isCustom: false,
    type: 'perf',
};
export var initMetric = function (name, value) {
    return __assign({ name: name, value: value }, defaultMetricContext);
};
export var invokeGetOnce = function (cb) {
    var hasGetted = false;
    var invoke = function (params) {
        if (hasGetted)
            return undefined;
        // 如果执行后依旧未拿到，直接返回
        var value = cb && cb(params);
        if (!value)
            return undefined;
        // 标记【拿到过数据】并返回数据
        hasGetted = true;
        return value;
    };
    return [invoke];
};
export var buildSingleMetricPayload = function (singleMetric) { return ({
    ev_type: PERFORMANCE_METRIC_EV_TYPE,
    payload: singleMetric,
}); };
export var onceHidden = function (cb, once) {
    if (once === void 0) { once = true; }
    var document = getDefaultDocument();
    if (!document)
        return;
    var removeListener = noop;
    if (document.visibilityState === 'hidden') {
        cb();
        return;
    }
    var onVisibilityChange = function () {
        if (document.visibilityState === 'hidden') {
            cb();
            once && removeListener();
        }
    };
    removeListener = addListenerToDocument(document, 'visibilitychange', onVisibilityChange, true);
};//# sourceMappingURL=https://trae.private/sourcemaps/424b4bd987c6c6a4cadf1a08da420593cc6bf75a/node_modules/@byted-icube/devtools-base/node_modules/@slardar/sdk-web/esm/monitors/performance/util.js.map
