// @slardar/sdk-template/esm/plugins/__test__/common.js
import { createClient } from '../../client';
import { createBaseConfigManager, id } from '../../utils';
export var createTestClient = function (send) {
    return createClient({
        builder: {
            build: function (ev) {
                var ev_type = ev.ev_type, payload = ev.payload, extra = ev.extra;
                var _a = extra || {}, _b = _a.bid, bid = _b === void 0 ? '' : _b, _d = _a.url, url = _d === void 0 ? '' : _d;
                return {
                    ev_type: ev_type,
                    payload: payload,
                    common: {
                        bid: bid,
                        url: url,
                    },
                };
            },
        },
        createSender: function () { return ({
            send: send,
        }); },
        createDefaultConfig: function (_c) { return ({ bid: '', url: '' }); },
        userConfigNormalizer: id,
        initConfigNormalizer: id,
        createConfigManager: createBaseConfigManager,
        validateInitConfig: function (_c) { return true; },
    });
};//# sourceMappingURL=https://trae.private/sourcemaps/424b4bd987c6c6a4cadf1a08da420593cc6bf75a/node_modules/@slardar/sdk-template/esm/plugins/__test__/common.js.map
