// @ai-local-server/common/dist/common/constant.js
import { Product, ProductAndRegion, Region } from '../typings';
import { AIServerRegion } from './enum';
export const SETTINGS_SERVICE_URL_MAP = {
    [Region.CN]: `https://is.snssdk.com/service/settings/v3/`,
    [Region.BOE]: `https://is.snssdk.com/service/settings/v3/`,
    [Region.BOEI18N]: `https://is.snssdk.com/service/settings/v3/`,
    [Region.SG]: `https://is.snssdk.com/service/settings/v3/`,
    [Region.US]: `https://is.snssdk.com/service/settings/v3/`,
};
export const SETTINGS_INTERVAL_TIME = 1000 * 60 * 5;
export const API_SERVER_INTERVAL_TIME = 1000 * 60 * 5;
export const STARLING_ZONE_HOST_MAP = {
    [Region.CN]: 'https://starling.zijieapi.com',
    [Region.BOE]: 'https://starling.zijieapi.com',
    [Region.US]: 'https://starling-oversea.byteoversea.com',
    [Region.SG]: 'https://starling-oversea.byteoversea.com',
    [Region.BOEI18N]: 'https://starling-oversea.byteoversea.com',
};
// https://bytedance.larkoffice.com/wiki/FqfMwibDBi3SkNk1fYmcaLZanxb#BqzldwyrRoRlZHxwop7cQa6Fnxg
export const AI_REQUEST_HOST_MAP = {
    // 对于 marscode 来说，boe 和 boei18n 域名跟线上是一致的
    [ProductAndRegion.MarsCodeBOE]: 'https://a0ai-api.zijieapi.com',
    [ProductAndRegion.MarsCodeBOEI18N]: 'https://a0ai-api.byteintlapi.com',
    [ProductAndRegion.MarsCodeUS]: 'https://a0ai-api.byteintlapi.com',
    [ProductAndRegion.MarsCodeSG]: 'https://a0ai-api.byteintlapi.com',
    [ProductAndRegion.MarsCodeCN]: 'https://a0ai-api.zijieapi.com',
    [ProductAndRegion.CloudIDEBOE]: 'https://copilot.byted.org',
    [ProductAndRegion.CloudIDECN]: 'https://copilot.byted.org',
    [ProductAndRegion.CloudIDEUS]: 'https://copilot.byteintl.net',
    [ProductAndRegion.CloudIDESG]: 'https://copilot-sg.byteintl.net',
    [ProductAndRegion.CloudIDEBOEI18N]: 'https://copilot.byteintl.net',
    [ProductAndRegion.CozePluginBOE]: 'https://a0ai-api.zijieapi.com',
    [ProductAndRegion.CozePluginCN]: 'https://a0ai-api.zijieapi.com',
    [ProductAndRegion.MerlinBOE]: 'https://copilot.byted.org',
    [ProductAndRegion.MerlinCN]: 'https://copilot.byted.org',
    [ProductAndRegion.PracticeBOE]: 'https://a0ai-api.zijieapi.com',
    [ProductAndRegion.PracticeCN]: 'https://a0ai-api.zijieapi.com',
    [ProductAndRegion.MarsCodeInternalCN]: 'https://copilot.byted.org',
    [ProductAndRegion.MarsCodeInternalBOE]: 'https://copilot.byted.org',
    [ProductAndRegion.DesktopInternalCN]: 'https://copilot.byted.org',
    [ProductAndRegion.DesktopInternalSG]: 'https://copilot-sg-og.byteintl.net',
    [ProductAndRegion.DesktopInternalUS]: 'https://copilot-og.byteintl.net',
    [ProductAndRegion.CourseBOE]: 'https://a0ai-api.zijieapi.com',
    [ProductAndRegion.CourseCN]: 'https://a0ai-api.zijieapi.com',
};
export const DEV_AI_REQUEST_HOST_MAP = {
    [ProductAndRegion.CloudIDEUS]: 'https://copilot-og.byteintl.net',
    [ProductAndRegion.CloudIDESG]: 'https://copilot-sg-og.byteintl.net',
};
export var RequestPathType;
(function (RequestPathType) {
    RequestPathType["DOMAIN_CONFIG"] = "api/ide/v1/domains_config";
    RequestPathType["PRACTICE_TITLE"] = "api/ide/v1/practice/generate_conversation_title";
    RequestPathType["PRACTICE_TEST_CASE"] = "/api/ide/v1/practice/generate_test_case";
    RequestPathType["CHAT"] = "api/ide/v1/chat";
    RequestPathType["INTENT_DETECT"] = "api/ide/v1/intent_detect";
    RequestPathType["INTENT_CONFIG"] = "api/ide/v1/intents_config";
    RequestPathType["FEEDBACK"] = "api/ide/v1/feedback";
    RequestPathType["APP_VERSIONS"] = "api/ide/v1/app_versions";
    RequestPathType["FEATURES"] = "api/ide/v1/features";
    RequestPathType["APPLY_CODE_SNIPPT"] = "api/ide/v1/fast_apply";
})(RequestPathType || (RequestPathType = {}));
export const REQUEST_AUTH_HEADER_MAP = {
    [Product.CozePlugin]: 'authorization: Cloud-IDE-JWT $token',
    [Product.CloudIDE]: 'x-jwt-token: $token',
    [Product.AITerminal]: 'x-jwt-token: $token',
    [Product.MarsCode]: 'x-ide-token: $token',
    [Product.Practice]: 'x-ide-token: $token',
    [Product.Merlin]: 'x-jwt-token: $token',
    [Product.MarsCodeInternal]: 'x-jwt-token: $token',
    [Product.DesktopInternal]: 'x-ide-token: $token',
    [Product.Course]: 'x-ide-token: $token',
};
export const AI_SEVER_REGION_MAP = {
    [Region.BOE]: AIServerRegion.CN,
    [Region.BOEI18N]: AIServerRegion.US,
    [Region.CN]: AIServerRegion.CN,
    [Region.SG]: AIServerRegion.SG,
    [Region.US]: AIServerRegion.US,
};
// https://bytedance.larkoffice.com/wiki/FqfMwibDBi3SkNk1fYmcaLZanxb#NVDFdoYkMosmmfx5gq4cXOQpn1d
export const AI_SERVER_APP_ID = {
    [Product.CloudIDE]: 'b070098d-978d-4274-ae6b-b76da2570112',
    [Product.CozePlugin]: '300ac202-684e-42c4-90a8-192a48aa35c1',
    [Product.MarsCode]: 'b070098d-978d-4274-ae6b-b76da2570112',
    [Product.Practice]: '850edec7-b9d0-48aa-99b5-67c888e282cd',
    [Product.Merlin]: '300ac202-684e-42c4-90a8-192a48aa35c1',
    [Product.MarsCodeInternal]: '85132071-b1fb-4f56-b532-77b1d7b6e81e',
    [Product.DesktopInternal]: '6eefa01c-1036-4c7e-9ca5-d891f63bfcd8',
    [Product.Course]: 'b070098d-978d-4274-ae6b-b76da2570112',
    [Product.AITerminal]: 'dadc14f9-2dbf-4dda-8e24-13d26096670d',
};
export const SUPPORTED_REGION_FOR_PRODUCT = {
    [Product.CloudIDE]: [Region.CN, Region.BOE, Region.BOEI18N, Region.SG, Region.US],
    [Product.AITerminal]: [Region.CN, Region.BOE, Region.BOEI18N, Region.SG, Region.US],
    [Product.MarsCode]: [Region.CN, Region.BOE, Region.BOEI18N, Region.SG, Region.US],
    [Product.CozePlugin]: [Region.CN, Region.BOE],
    [Product.Merlin]: [Region.CN, Region.BOE],
    [Product.Practice]: [Region.CN, Region.BOE],
    [Product.MarsCodeInternal]: [Region.CN, Region.BOE],
    [Product.DesktopInternal]: [Region.CN, Region.SG, Region.US],
    [Product.Course]: [Region.CN, Region.BOE],
};
export const PRODUCT_AND_REGION_MAP = {
    [Product.CloudIDE]: {
        [Region.CN]: ProductAndRegion.CloudIDECN,
        [Region.BOE]: ProductAndRegion.CloudIDEBOE,
        [Region.SG]: ProductAndRegion.CloudIDESG,
        [Region.US]: ProductAndRegion.CloudIDEUS,
        [Region.BOEI18N]: ProductAndRegion.CloudIDEBOEI18N,
    },
    [Product.AITerminal]: {
        [Region.CN]: ProductAndRegion.CloudIDECN,
        [Region.BOE]: ProductAndRegion.CloudIDEBOE,
        [Region.SG]: ProductAndRegion.CloudIDESG,
        [Region.US]: ProductAndRegion.CloudIDEUS,
        [Region.BOEI18N]: ProductAndRegion.CloudIDEBOEI18N,
    },
    [Product.MarsCode]: {
        [Region.CN]: ProductAndRegion.MarsCodeCN,
        [Region.BOE]: ProductAndRegion.MarsCodeBOE,
        [Region.SG]: ProductAndRegion.MarsCodeSG,
        [Region.BOEI18N]: ProductAndRegion.MarsCodeBOEI18N,
        [Region.US]: ProductAndRegion.MarsCodeUS,
    },
    [Product.CozePlugin]: {
        [Region.CN]: ProductAndRegion.CozePluginCN,
        [Region.BOE]: ProductAndRegion.CozePluginBOE,
    },
    [Product.Merlin]: {
        [Region.CN]: ProductAndRegion.MerlinCN,
        [Region.BOE]: ProductAndRegion.MerlinBOE,
    },
    [Product.Practice]: {
        [Region.CN]: ProductAndRegion.PracticeCN,
        [Region.BOE]: ProductAndRegion.PracticeBOE,
    },
    [Product.MarsCodeInternal]: {
        [Region.CN]: ProductAndRegion.MarsCodeInternalCN,
        [Region.BOE]: ProductAndRegion.MarsCodeInternalBOE,
    },
    [Product.DesktopInternal]: {
        [Region.CN]: ProductAndRegion.DesktopInternalCN,
        [Region.SG]: ProductAndRegion.DesktopInternalSG,
        [Region.US]: ProductAndRegion.DesktopInternalUS,
    },
    [Product.Course]: {
        [Region.CN]: ProductAndRegion.CourseCN,
        [Region.BOE]: ProductAndRegion.CourseBOE,
    },
};
export const SLARDAR_BID_MAP = {
    [ProductAndRegion.CloudIDECN]: 'icube_web',
    [ProductAndRegion.CloudIDEBOE]: 'icube_web',
    [ProductAndRegion.CloudIDESG]: 'cloud_ide_us',
    [ProductAndRegion.CloudIDEUS]: 'cloud_ide_us',
    [ProductAndRegion.CloudIDEBOEI18N]: 'cloud_ide_us',
    [ProductAndRegion.MarsCodeCN]: 'stable_ide_cn',
    [ProductAndRegion.MarsCodeBOE]: 'stable_ide_cn',
    [ProductAndRegion.MarsCodeSG]: 'stable_ide_us',
    [ProductAndRegion.MarsCodeUS]: 'stable_ide_us',
    [ProductAndRegion.MarsCodeBOEI18N]: 'stable_ide_us',
    [ProductAndRegion.CozePluginCN]: 'icube_web',
    [ProductAndRegion.CozePluginBOE]: 'icube_web',
    [ProductAndRegion.MerlinCN]: 'icube_web',
    [ProductAndRegion.MerlinBOE]: 'icube_web',
    [ProductAndRegion.PracticeCN]: 'stable_ide_cn',
    [ProductAndRegion.PracticeBOE]: 'stable_ide_cn',
    [ProductAndRegion.MarsCodeInternalCN]: 'stable_ide_cn',
    [ProductAndRegion.MarsCodeInternalBOE]: 'stable_ide_cn',
    [ProductAndRegion.DesktopInternalCN]: 'stable_ide_cn',
    [ProductAndRegion.DesktopInternalUS]: 'stable_ide_us',
    [ProductAndRegion.DesktopInternalSG]: 'stable_ide_us ',
    [ProductAndRegion.CourseCN]: 'stable_ide_cn',
    [ProductAndRegion.CourseBOE]: 'stable_ide_cn',
};
