// @slardar/sdk-web/esm/plugins/monitors/breadcrumb.js
import { __read } from "tslib";
import { initSubjectInGlobal, registerTearDownGroup } from '@slardar/sdk-template';
import { clickSubject, keypressSubject } from '../../collector/dom';
import { HTTP_EV_TYPE, JS_ERROR_EV_TYPE } from '../../constants';
import { breadcrumbGetterWithDomObserver, } from '../../monitors/breadcrumb';
import { getConfig } from '../../utils';
export var BREADCRUMB_MONITOR_PLUGIN_NAME = 'breadcrumb';
var defaultConfig = {
    maxBreadcrumbs: 20,
    dom: true,
};
export function BreadcrumbMonitorBasePlugin(client, c) {
    var config = getConfig(c, defaultConfig);
    if (!config) {
        return;
    }
    var tearDownGroup = [];
    var _a = __read(breadcrumbGetterWithDomObserver(tearDownGroup, [initSubjectInGlobal(client, clickSubject), initSubjectInGlobal(client, keypressSubject)], config), 2), getBreadcrumbs = _a[0], addBreadcrumb = _a[1];
    var listenOtherReport = function (ev) {
        if (ev.ev_type === HTTP_EV_TYPE) {
            addBreadcrumb({
                type: HTTP_EV_TYPE,
                category: ev.payload.api,
                message: '',
                data: {
                    method: ev.payload.request.method,
                    url: ev.payload.request.url,
                    status_code: String(ev.payload.response.status),
                },
                timestamp: ev.payload.request.timestamp,
            });
        }
        return ev;
    };
    client.on('report', listenOtherReport);
    tearDownGroup.push(function () {
        client.off('report', listenOtherReport);
    });
    registerTearDownGroup(client, BREADCRUMB_MONITOR_PLUGIN_NAME, JS_ERROR_EV_TYPE, tearDownGroup);
    client.provide('getBreadcrumbs', getBreadcrumbs);
    client.provide('addBreadcrumb', addBreadcrumb);
}
export function BreadcrumbMonitorPlugin(client) {
    client.on('init', function () {
        var _a;
        var config = (_a = client.config()) === null || _a === void 0 ? void 0 : _a.plugins[BREADCRUMB_MONITOR_PLUGIN_NAME];
        BreadcrumbMonitorBasePlugin(client, config);
    });
}//# sourceMappingURL=https://trae.private/sourcemaps/424b4bd987c6c6a4cadf1a08da420593cc6bf75a/node_modules/@slardar/base/node_modules/@slardar/sdk-web/esm/plugins/monitors/breadcrumb.js.map
