// @slardar/sdk-web/esm/utils/__test__/util.js
import { createBaseConfigManager, createClient, directBuilder, id, noopSender, SubjectPlugin, } from '@slardar/sdk-template';
import { now } from '../date';
export var createTestWebClient = function () {
    var client = createClient({
        builder: directBuilder,
        createSender: function () { return noopSender; },
        createDefaultConfig: function () { return ({
            pid: '',
            viewId: '_' + now(),
            userId: '',
            deviceId: '',
            sessionId: '',
            plugins: {},
            sample: {
                sample_rate: 1,
                sample_granularity: 'session',
                include_users: [],
                rules: {},
            },
        }); },
        userConfigNormalizer: id,
        initConfigNormalizer: id,
        createConfigManager: createBaseConfigManager,
        validateInitConfig: function () { return true; },
    });
    SubjectPlugin(client, undefined);
    return client;
};//# sourceMappingURL=https://trae.private/sourcemaps/424b4bd987c6c6a4cadf1a08da420593cc6bf75a/node_modules/@byted-icube/devtools-base/node_modules/@slardar/sdk-web/esm/utils/__test__/util.js.map
