// @byted-icube/devtools-base/dist/types/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Tenant = exports.ApiServerMap = exports.RegionMap = exports.Region = void 0;
var Region;
(function (Region) {
    Region["CN"] = "CN";
    Region["BOE"] = "BOE";
    Region["SG"] = "SG";
    Region["US"] = "US";
    Region["BOEI18N"] = "BOEI18N";
})(Region || (exports.Region = Region = {}));
exports.RegionMap = {
    [Region.CN]: 'cn',
    [Region.BOE]: 'cn',
    [Region.SG]: 'sg',
    [Region.US]: 'us',
    [Region.BOEI18N]: 'sg',
};
exports.ApiServerMap = {
    [Region.CN]: 'https://api.marscode.cn',
    [Region.BOE]: 'https://bytesec.bytedance.com',
    [Region.SG]: 'https://api-sg-central.marscode.com',
    [Region.US]: 'https://api-sg-central.marscode.com',
    [Region.BOEI18N]: 'https://bytesec.byteintlapi.com',
};
var Tenant;
(function (Tenant) {
    Tenant["BYTEDANCE"] = "bytedance";
    Tenant["PUBLIC"] = "public";
    Tenant["UNKNOWN"] = "unknown";
})(Tenant || (exports.Tenant = Tenant = {}));
