// @ai-local-server/common/dist/typings/http.js
export var ERRCODE;
(function (ERRCODE) {
    ERRCODE[ERRCODE["DEFAULT"] = 0] = "DEFAULT";
    ERRCODE[ERRCODE["NO_WORKSPACE"] = -1] = "NO_WORKSPACE";
    ERRCODE[ERRCODE["ETIMEDOUT"] = -2] = "ETIMEDOUT";
    ERRCODE[ERRCODE["ECONNABORTED"] = -3] = "ECONNABORTED";
    ERRCODE[ERRCODE["ENOTFOUND"] = -4] = "ENOTFOUND";
    ERRCODE[ERRCODE["ECONNRESET"] = -5] = "ECONNRESET";
    ERRCODE[ERRCODE["CERT_HAS_EXPIRED"] = -6] = "CERT_HAS_EXPIRED";
    ERRCODE[ERRCODE["ERR_BAD_REQUEST"] = -7] = "ERR_BAD_REQUEST";
    ERRCODE[ERRCODE["INVALID_TOKEN"] = 1001] = "INVALID_TOKEN";
    ERRCODE[ERRCODE["AUTH_FAIL"] = 1002] = "AUTH_FAIL";
    ERRCODE[ERRCODE["UNKNOWN_ERROR"] = 2001] = "UNKNOWN_ERROR";
    ERRCODE[ERRCODE["MODEL_FAIL"] = 3003] = "MODEL_FAIL";
    ERRCODE[ERRCODE["REQUEST_LIMIT"] = 3004] = "REQUEST_LIMIT";
    ERRCODE[ERRCODE["INVALID_ARGS"] = 4001] = "INVALID_ARGS";
    ERRCODE[ERRCODE["TOKEN_LIMIT"] = 4006] = "TOKEN_LIMIT";
    ERRCODE[ERRCODE["USER_REQUEST_LIMIT"] = 4007] = "USER_REQUEST_LIMIT";
    ERRCODE[ERRCODE["ABNORMAL_ACCOUNT_LOGOUT"] = 4010] = "ABNORMAL_ACCOUNT_LOGOUT";
    ERRCODE[ERRCODE["ABNORMAL_ACCOUNT_LIMIT"] = 4011] = "ABNORMAL_ACCOUNT_LIMIT";
    ERRCODE[ERRCODE["REQUEST_DOWNGRADE"] = 4012] = "REQUEST_DOWNGRADE";
    ERRCODE[ERRCODE["REQUEST_ABNORMAL"] = 4013] = "REQUEST_ABNORMAL";
    ERRCODE[ERRCODE["TODAY_USER_REQUEST_LIMIT"] = 4021] = "TODAY_USER_REQUEST_LIMIT";
    ERRCODE[ERRCODE["TURN_EXCEEDED"] = 5001] = "TURN_EXCEEDED";
    ERRCODE[ERRCODE["TIMEOUT"] = 5002] = "TIMEOUT";
    ERRCODE[ERRCODE["OUT_OF_QUOTA"] = 5003] = "OUT_OF_QUOTA";
    ERRCODE[ERRCODE["AGENT_BUSY"] = 5004] = "AGENT_BUSY";
})(ERRCODE || (ERRCODE = {}));
export var ERROR_LEVEL;
(function (ERROR_LEVEL) {
    ERROR_LEVEL["WARN"] = "warn";
    ERROR_LEVEL["ERROR"] = "error";
})(ERROR_LEVEL || (ERROR_LEVEL = {}));
export const ERROR_CODE_MAP = {
    [ERRCODE.DEFAULT]: {
        key: 'icube.error.default',
        level: ERROR_LEVEL.ERROR,
    },
    [ERRCODE.NO_WORKSPACE]: {
        key: 'icube.error.noWorkspace',
        level: ERROR_LEVEL.WARN,
    },
    [ERRCODE.ETIMEDOUT]: {
        key: 'icube.error.eTimeout',
        level: ERROR_LEVEL.ERROR,
    },
    [ERRCODE.ECONNABORTED]: {
        key: 'icube.error.eConnAborted',
        level: ERROR_LEVEL.ERROR,
    },
    [ERRCODE.ENOTFOUND]: {
        key: 'icube.error.eNotFound',
        level: ERROR_LEVEL.ERROR,
    },
    [ERRCODE.ECONNRESET]: {
        key: 'icube.error.eConnReset',
        level: ERROR_LEVEL.ERROR,
    },
    [ERRCODE.CERT_HAS_EXPIRED]: {
        key: 'icube.error.certHasExpired',
        level: ERROR_LEVEL.ERROR,
    },
    [ERRCODE.ERR_BAD_REQUEST]: {
        key: 'icube.error.eBadRequest',
        level: ERROR_LEVEL.ERROR,
    },
    [ERRCODE.INVALID_TOKEN]: {
        key: 'icube.error.invalidToken',
        level: ERROR_LEVEL.WARN,
    },
    [ERRCODE.AUTH_FAIL]: {
        key: 'icube.error.authFail',
        level: ERROR_LEVEL.WARN,
    },
    [ERRCODE.UNKNOWN_ERROR]: {
        key: 'icube.error.unknownError',
        level: ERROR_LEVEL.ERROR,
    },
    [ERRCODE.MODEL_FAIL]: {
        key: 'icube.error.modelFail',
        level: ERROR_LEVEL.ERROR,
    },
    [ERRCODE.REQUEST_LIMIT]: {
        key: 'icube.error.requestLimit',
        level: ERROR_LEVEL.WARN,
    },
    [ERRCODE.INVALID_ARGS]: {
        key: 'icube.error.invalidArgs',
        level: ERROR_LEVEL.WARN,
    },
    [ERRCODE.TOKEN_LIMIT]: {
        key: 'icube.error.tokenLimit',
        level: ERROR_LEVEL.WARN,
    },
    [ERRCODE.USER_REQUEST_LIMIT]: {
        key: 'icube.error.userRequestLimit',
        level: ERROR_LEVEL.WARN,
    },
    [ERRCODE.ABNORMAL_ACCOUNT_LOGOUT]: {
        key: 'icube.error.abnormalAccountLogout',
        level: ERROR_LEVEL.WARN,
    },
    [ERRCODE.ABNORMAL_ACCOUNT_LIMIT]: {
        key: 'icube.error.abnormalAccountLimit',
        level: ERROR_LEVEL.WARN,
    },
    [ERRCODE.REQUEST_DOWNGRADE]: {
        key: 'icube.error.requestDowngrade',
        level: ERROR_LEVEL.WARN,
    },
    [ERRCODE.REQUEST_ABNORMAL]: {
        key: 'icube.error.requestAbnormal',
        level: ERROR_LEVEL.WARN,
    },
    [ERRCODE.TODAY_USER_REQUEST_LIMIT]: {
        key: 'icube.error.todayUserRequestLimit',
        level: ERROR_LEVEL.WARN,
    },
    [ERRCODE.TURN_EXCEEDED]: {
        key: 'icube.error.turnExceeded',
        level: ERROR_LEVEL.WARN,
    },
    [ERRCODE.TIMEOUT]: {
        key: 'icube.error.timeout',
        level: ERROR_LEVEL.ERROR,
    },
    [ERRCODE.OUT_OF_QUOTA]: {
        key: 'icube.error.outOfQuota',
        level: ERROR_LEVEL.WARN,
    },
    [ERRCODE.AGENT_BUSY]: {
        key: 'icube.error.agentBusy',
        level: ERROR_LEVEL.WARN,
    },
};
