// @ai-local-server/common/dist/typings/slardar.js
export var SlardarEventStatus;
(function (SlardarEventStatus) {
    SlardarEventStatus["Success"] = "Success";
    SlardarEventStatus["Failed"] = "Failed";
    SlardarEventStatus["None"] = "None";
})(SlardarEventStatus || (SlardarEventStatus = {}));
export var SlardarEventName;
(function (SlardarEventName) {
    SlardarEventName["AIExtensionActive"] = "icube_ai_extension_active";
    SlardarEventName["StartCkgServer"] = "icube_ai_start_ckg_server";
    SlardarEventName["QueryIntent"] = "icube_ai_query_intent";
    SlardarEventName["ContextResolve"] = "icube_ai_context_resolve";
    SlardarEventName["BuildPrompt"] = "icube_ai_build_prompt";
    SlardarEventName["FirstToken"] = "icube_ai_chat_first_response_token";
    SlardarEventName["ChatResponse"] = "icube_ai_chat_response";
    SlardarEventName["InitSession"] = "icube_ai_init_session";
    SlardarEventName["ChatRequest"] = "icube_ai_chat_request";
    SlardarEventName["InlineChatRequest"] = "icube_ai_inline_chat_request";
    SlardarEventName["AIMentionFileSearch"] = "icube_ai_mention_file_search";
    SlardarEventName["AIMentionCodeSearch"] = "icube_ai_mention_code_search";
    SlardarEventName["GlobalChatRequest"] = "icube_ai_global_chat_request";
    SlardarEventName["UnhandleRejection"] = "icube_ai_unhandle_rejection";
    SlardarEventName["UncaughtException"] = "icube_ai_uncaught_exception";
    SlardarEventName["CKGResolver"] = "icube_ai_ckg_resolver";
    SlardarEventName["CKGExit"] = "icube_ai_ckg_exit";
})(SlardarEventName || (SlardarEventName = {}));
