// @slardar/integrations/dist/modifyPageviewContext.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var noop = function () { return ({}); };

/* eslint-disable @typescript-eslint/prefer-nullish-coalescing */
var MODIFY_PAGEVIEW_CONTEXT_INTEGRATION_NAME = 'modifyPageviewContext';
var modifyPageviewContext = function () {
    var tearDownFn = noop;
    return {
        name: MODIFY_PAGEVIEW_CONTEXT_INTEGRATION_NAME,
        setup: function (client) {
            var onReport = function (ev) {
                if (ev.ev_type === 'pageview') {
                    var config = client.config();
                    var currentPid = ev.payload.pid;
                    config.pid !== currentPid &&
                        client.set({
                            pid: currentPid,
                            viewId: currentPid + '_' + Date.now(),
                        });
                }
                return ev;
            };
            client.on('report', onReport);
            tearDownFn = function () { return client.off('report', onReport); };
        },
        tearDown: function () { return tearDownFn(); },
    };
};

exports.MODIFY_PAGEVIEW_CONTEXT_INTEGRATION_NAME = MODIFY_PAGEVIEW_CONTEXT_INTEGRATION_NAME;
exports.modifyPageviewContext = modifyPageviewContext;//# sourceMappingURL=https://trae.private/sourcemaps/424b4bd987c6c6a4cadf1a08da420593cc6bf75a/node_modules/@byted-icube/devtools-base/node_modules/@slardar/integrations/dist/modifyPageviewContext.js.map
