// @slardar/integrations/dist/refreshPreStartContext.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

/*! *****************************************************************************
Copyright (c) Microsoft Corporation.

Permission to use, copy, modify, and/or distribute this software for any
purpose with or without fee is hereby granted.

THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES WITH
REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY
AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT,
INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM
LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR
OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
PERFORMANCE OF THIS SOFTWARE.
***************************************************************************** */

var __assign = function() {
    __assign = Object.assign || function __assign(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p)) t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};

var noop = function () { return ({}); };

/*
 * @Author: pengli.coder
 * @Date: 2021-10-27 20:04:51
 * @LastEditTime: 2021-10-27 20:30:06
 * @LastEditors: Please set LastEditors
 * @Description: refresh the context of the pre-start queue with the context at start
 */
var REFRESH_PRE_START_CONTEXT_INTEGRATION_NAME = 'refreshPreStartContext';
var refreshPreStartContext = function () {
    var tearDownFn = noop;
    return {
        name: REFRESH_PRE_START_CONTEXT_INTEGRATION_NAME,
        setup: function (client) {
            var onStart = function () {
                var preStartQueue = client.getPreStartQueue();
                preStartQueue.forEach(function (ev, index) {
                    var _a, _b;
                    preStartQueue[index] = __assign(__assign({}, ev), { overrides: __assign(__assign({}, ((_a = ev.overrides) !== null && _a !== void 0 ? _a : {})), { 
                            // eslint-disable-next-line @typescript-eslint/prefer-nullish-coalescing
                            context: ((_b = client === null || client === void 0 ? void 0 : client.context) === null || _b === void 0 ? void 0 : _b.toString()) || {} }) });
                });
            };
            client.on('start', onStart);
            tearDownFn = function () { return client.off('start', onStart); };
        },
        tearDown: function () { return tearDownFn(); },
    };
};

exports.REFRESH_PRE_START_CONTEXT_INTEGRATION_NAME = REFRESH_PRE_START_CONTEXT_INTEGRATION_NAME;
exports.refreshPreStartContext = refreshPreStartContext;//# sourceMappingURL=https://trae.private/sourcemaps/424b4bd987c6c6a4cadf1a08da420593cc6bf75a/node_modules/@byted-icube/devtools-base/node_modules/@slardar/integrations/dist/refreshPreStartContext.js.map
