// @slardar/base/esm/__test__/utils.js
import { __assign } from "tslib";
import sinon from 'sinon';
function request(_a) {
    var fail = _a.fail, success = _a.success, $mock_res = _a.$mock_res;
    var DEFAULT_RES = {
        status: 200,
        headers: null,
        data: {
            data: {},
        },
        duration: 100,
    };
    var res = Object.assign(DEFAULT_RES, $mock_res);
    setTimeout(function () {
        if (res.error) {
            fail === null || fail === void 0 ? void 0 : fail(new Error(res.error));
        }
        else {
            success === null || success === void 0 ? void 0 : success(res);
        }
    }, res.duration);
}
export function getTransport() {
    return {
        get: sinon.spy(function (options) { return request(__assign({ method: 'GET' }, options)); }),
        post: sinon.spy(function (options) { return request(__assign({ method: 'POST' }, options)); }),
    };
}//# sourceMappingURL=https://trae.private/sourcemaps/424b4bd987c6c6a4cadf1a08da420593cc6bf75a/node_modules/@byted-icube/devtools-base/node_modules/@slardar/base/esm/__test__/utils.js.map
