// @slardar/web/oci/plugins/heatmap-draw.js
!(function () {
  'use strict'
  var e = function () {
    return (
      (e =
        Object.assign ||
        function (e) {
          for (var t, n = 1, r = arguments.length; n < r; n++)
            for (var o in (t = arguments[n])) Object.prototype.hasOwnProperty.call(t, o) && (e[o] = t[o])
          return e
        }),
      e.apply(this, arguments)
    )
  }
  function t(e, t, n, r) {
    return new (n || (n = Promise))(function (o, i) {
      function a(e) {
        try {
          u(r.next(e))
        } catch (e) {
          i(e)
        }
      }
      function s(e) {
        try {
          u(r.throw(e))
        } catch (e) {
          i(e)
        }
      }
      function u(e) {
        var t
        e.done
          ? o(e.value)
          : ((t = e.value),
            t instanceof n
              ? t
              : new n(function (e) {
                  e(t)
                })).then(a, s)
      }
      u((r = r.apply(e, t || [])).next())
    })
  }
  function n(e, t) {
    var n,
      r,
      o,
      i,
      a = {
        label: 0,
        sent: function () {
          if (1 & o[0]) throw o[1]
          return o[1]
        },
        trys: [],
        ops: [],
      }
    return (
      (i = { next: s(0), throw: s(1), return: s(2) }),
      'function' == typeof Symbol &&
        (i[Symbol.iterator] = function () {
          return this
        }),
      i
    )
    function s(i) {
      return function (s) {
        return (function (i) {
          if (n) throw new TypeError('Generator is already executing.')
          for (; a; )
            try {
              if (
                ((n = 1),
                r &&
                  (o = 2 & i[0] ? r.return : i[0] ? r.throw || ((o = r.return) && o.call(r), 0) : r.next) &&
                  !(o = o.call(r, i[1])).done)
              )
                return o
              switch (((r = 0), o && (i = [2 & i[0], o.value]), i[0])) {
                case 0:
                case 1:
                  o = i
                  break
                case 4:
                  return a.label++, { value: i[1], done: !1 }
                case 5:
                  a.label++, (r = i[1]), (i = [0])
                  continue
                case 7:
                  ;(i = a.ops.pop()), a.trys.pop()
                  continue
                default:
                  if (!((o = a.trys), (o = o.length > 0 && o[o.length - 1]) || (6 !== i[0] && 2 !== i[0]))) {
                    a = 0
                    continue
                  }
                  if (3 === i[0] && (!o || (i[1] > o[0] && i[1] < o[3]))) {
                    a.label = i[1]
                    break
                  }
                  if (6 === i[0] && a.label < o[1]) {
                    ;(a.label = o[1]), (o = i)
                    break
                  }
                  if (o && a.label < o[2]) {
                    ;(a.label = o[2]), a.ops.push(i)
                    break
                  }
                  o[2] && a.ops.pop(), a.trys.pop()
                  continue
              }
              i = t.call(e, a)
            } catch (e) {
              ;(i = [6, e]), (r = 0)
            } finally {
              n = o = 0
            }
          if (5 & i[0]) throw i[1]
          return { value: i[0] ? i[1] : void 0, done: !0 }
        })([i, s])
      }
    }
  }
  function r(e, t, n) {
    if (n || 2 === arguments.length)
      for (var r, o = 0, i = t.length; o < i; o++)
        (!r && o in t) || (r || (r = Array.prototype.slice.call(t, 0, o)), (r[o] = t[o]))
    return e.concat(r || Array.prototype.slice.call(t))
  }
  var o,
    i,
    a,
    s =
      'undefined' != typeof globalThis
        ? globalThis
        : 'undefined' != typeof window
        ? window
        : 'undefined' != typeof global
        ? global
        : 'undefined' != typeof self
        ? self
        : {},
    u = { exports: {} }
  ;(i = s),
    (a = function () {
      var e,
        t = {
          defaultRadius: 40,
          defaultRenderer: 'canvas2d',
          defaultGradient: { 0.25: 'rgb(0,0,255)', 0.55: 'rgb(0,255,0)', 0.85: 'yellow', 1: 'rgb(255,0,0)' },
          defaultMaxOpacity: 1,
          defaultMinOpacity: 0,
          defaultBlur: 0.85,
          defaultXField: 'x',
          defaultYField: 'y',
          defaultValueField: 'value',
          plugins: {},
        },
        n = (function () {
          var e = function (e) {
              ;(this._coordinator = {}),
                (this._data = []),
                (this._radi = []),
                (this._min = 10),
                (this._max = 1),
                (this._xField = e.xField || e.defaultXField),
                (this._yField = e.yField || e.defaultYField),
                (this._valueField = e.valueField || e.defaultValueField),
                e.radius && (this._cfgRadius = e.radius)
            },
            n = t.defaultRadius
          return (
            (e.prototype = {
              _organiseData: function (e, t) {
                var r = e[this._xField],
                  o = e[this._yField],
                  i = this._radi,
                  a = this._data,
                  s = this._max,
                  u = this._min,
                  c = e[this._valueField] || 1,
                  l = e.radius || this._cfgRadius || n
                a[r] || ((a[r] = []), (i[r] = [])), a[r][o] ? (a[r][o] += c) : ((a[r][o] = c), (i[r][o] = l))
                var f = a[r][o]
                return f > s
                  ? (t ? this.setDataMax(f) : (this._max = f), !1)
                  : f < u
                  ? (t ? this.setDataMin(f) : (this._min = f), !1)
                  : { x: r, y: o, value: c, radius: l, min: u, max: s }
              },
              _unOrganizeData: function () {
                var e = [],
                  t = this._data,
                  n = this._radi
                for (var r in t) for (var o in t[r]) e.push({ x: r, y: o, radius: n[r][o], value: t[r][o] })
                return { min: this._min, max: this._max, data: e }
              },
              _onExtremaChange: function () {
                this._coordinator.emit('extremachange', { min: this._min, max: this._max })
              },
              addData: function () {
                if (arguments[0].length > 0)
                  for (var e = arguments[0], t = e.length; t--; ) this.addData.call(this, e[t])
                else {
                  var n = this._organiseData(arguments[0], !0)
                  n &&
                    (0 === this._data.length && (this._min = this._max = n.value),
                    this._coordinator.emit('renderpartial', { min: this._min, max: this._max, data: [n] }))
                }
                return this
              },
              setData: function (e) {
                var t = e.data,
                  n = t.length
                ;(this._data = []), (this._radi = [])
                for (var r = 0; r < n; r++) this._organiseData(t[r], !1)
                return (
                  (this._max = e.max),
                  (this._min = e.min || 0),
                  this._onExtremaChange(),
                  this._coordinator.emit('renderall', this._getInternalData()),
                  this
                )
              },
              removeData: function () {},
              setDataMax: function (e) {
                return (
                  (this._max = e),
                  this._onExtremaChange(),
                  this._coordinator.emit('renderall', this._getInternalData()),
                  this
                )
              },
              setDataMin: function (e) {
                return (
                  (this._min = e),
                  this._onExtremaChange(),
                  this._coordinator.emit('renderall', this._getInternalData()),
                  this
                )
              },
              setCoordinator: function (e) {
                this._coordinator = e
              },
              _getInternalData: function () {
                return { max: this._max, min: this._min, data: this._data, radi: this._radi }
              },
              getData: function () {
                return this._unOrganizeData()
              },
            }),
            e
          )
        })(),
        r = (function () {
          var e = function (e) {
              var t = e.gradient || e.defaultGradient,
                n = document.createElement('canvas'),
                r = n.getContext('2d')
              ;(n.width = 256), (n.height = 1)
              var o = r.createLinearGradient(0, 0, 256, 1)
              for (var i in t) o.addColorStop(i, t[i])
              return (r.fillStyle = o), r.fillRect(0, 0, 256, 1), r.getImageData(0, 0, 256, 1).data
            },
            t = function (e, t) {
              var n = document.createElement('canvas'),
                r = n.getContext('2d'),
                o = e,
                i = e
              if (((n.width = n.height = 2 * e), 1 == t))
                r.beginPath(), r.arc(o, i, e, 0, 2 * Math.PI, !1), (r.fillStyle = 'rgba(0,0,0,1)'), r.fill()
              else {
                var a = r.createRadialGradient(o, i, e * t, o, i, e)
                a.addColorStop(0, 'rgba(0,0,0,1)'),
                  a.addColorStop(1, 'rgba(0,0,0,0)'),
                  (r.fillStyle = a),
                  r.fillRect(0, 0, 2 * e, 2 * e)
              }
              return n
            }
          function n(t) {
            var n = t.container,
              r = (this.shadowCanvas = document.createElement('canvas')),
              o = (this.canvas = t.canvas || document.createElement('canvas'))
            this._renderBoundaries = [1e4, 1e4, 0, 0]
            var i = getComputedStyle(t.container) || {}
            ;(o.className = 'heatmap-canvas'),
              (this._width = o.width = r.width = t.width || +i.width.replace(/px/, '')),
              (this._height = o.height = r.height = t.height || +i.height.replace(/px/, '')),
              (this.shadowCtx = r.getContext('2d')),
              (this.ctx = o.getContext('2d')),
              (o.style.cssText = r.style.cssText = 'position:absolute;left:0;top:0;'),
              (n.style.position = 'relative'),
              n.appendChild(o),
              (this._palette = e(t)),
              (this._templates = {}),
              this._setStyles(t)
          }
          return (
            (n.prototype = {
              renderPartial: function (e) {
                e.data.length > 0 && (this._drawAlpha(e), this._colorize())
              },
              renderAll: function (e) {
                this._clear(),
                  e.data.length > 0 &&
                    (this._drawAlpha(
                      (function (e) {
                        for (
                          var t = [],
                            n = e.min,
                            r = e.max,
                            o = e.radi,
                            i = ((e = e.data), Object.keys(e)),
                            a = i.length;
                          a--;

                        )
                          for (var s = i[a], u = Object.keys(e[s]), c = u.length; c--; ) {
                            var l = u[c],
                              f = e[s][l],
                              d = o[s][l]
                            t.push({ x: s, y: l, value: f, radius: d })
                          }
                        return { min: n, max: r, data: t }
                      })(e),
                    ),
                    this._colorize())
              },
              _updateGradient: function (t) {
                this._palette = e(t)
              },
              updateConfig: function (e) {
                e.gradient && this._updateGradient(e), this._setStyles(e)
              },
              setDimensions: function (e, t) {
                ;(this._width = e),
                  (this._height = t),
                  (this.canvas.width = this.shadowCanvas.width = e),
                  (this.canvas.height = this.shadowCanvas.height = t)
              },
              _clear: function () {
                this.shadowCtx.clearRect(0, 0, this._width, this._height),
                  this.ctx.clearRect(0, 0, this._width, this._height)
              },
              _setStyles: function (e) {
                ;(this._blur = 0 == e.blur ? 0 : e.blur || e.defaultBlur),
                  e.backgroundColor && (this.canvas.style.backgroundColor = e.backgroundColor),
                  (this._width = this.canvas.width = this.shadowCanvas.width = e.width || this._width),
                  (this._height = this.canvas.height = this.shadowCanvas.height = e.height || this._height),
                  (this._opacity = 255 * (e.opacity || 0)),
                  (this._maxOpacity = 255 * (e.maxOpacity || e.defaultMaxOpacity)),
                  (this._minOpacity = 255 * (e.minOpacity || e.defaultMinOpacity)),
                  (this._useGradientOpacity = !!e.useGradientOpacity)
              },
              _drawAlpha: function (e) {
                for (
                  var n = (this._min = e.min),
                    r = (this._max = e.max),
                    o = (e = e.data || []).length,
                    i = 1 - this._blur;
                  o--;

                ) {
                  var a,
                    s = e[o],
                    u = s.x,
                    c = s.y,
                    l = s.radius,
                    f = Math.min(s.value, r),
                    d = u - l,
                    p = c - l,
                    h = this.shadowCtx
                  this._templates[l] ? (a = this._templates[l]) : (this._templates[l] = a = t(l, i))
                  var m = (f - n) / (r - n)
                  ;(h.globalAlpha = m < 0.01 ? 0.01 : m),
                    h.drawImage(a, d, p),
                    d < this._renderBoundaries[0] && (this._renderBoundaries[0] = d),
                    p < this._renderBoundaries[1] && (this._renderBoundaries[1] = p),
                    d + 2 * l > this._renderBoundaries[2] && (this._renderBoundaries[2] = d + 2 * l),
                    p + 2 * l > this._renderBoundaries[3] && (this._renderBoundaries[3] = p + 2 * l)
                }
              },
              _colorize: function () {
                var e = this._renderBoundaries[0],
                  t = this._renderBoundaries[1],
                  n = this._renderBoundaries[2] - e,
                  r = this._renderBoundaries[3] - t,
                  o = this._width,
                  i = this._height,
                  a = this._opacity,
                  s = this._maxOpacity,
                  u = this._minOpacity,
                  c = this._useGradientOpacity
                e < 0 && (e = 0), t < 0 && (t = 0), e + n > o && (n = o - e), t + r > i && (r = i - t)
                for (
                  var l = this.shadowCtx.getImageData(e, t, n, r), f = l.data, d = f.length, p = this._palette, h = 3;
                  h < d;
                  h += 4
                ) {
                  var m,
                    v = f[h],
                    g = 4 * v
                  g &&
                    ((m = a > 0 ? a : v < s ? (v < u ? u : v) : s),
                    (f[h - 3] = p[g]),
                    (f[h - 2] = p[g + 1]),
                    (f[h - 1] = p[g + 2]),
                    (f[h] = c ? p[g + 3] : m))
                }
                this.ctx.putImageData(l, e, t), (this._renderBoundaries = [1e3, 1e3, 0, 0])
              },
              getValueAt: function (e) {
                var t = this.shadowCtx.getImageData(e.x, e.y, 1, 1).data[3],
                  n = this._max,
                  r = this._min
                return (Math.abs(n - r) * (t / 255)) >> 0
              },
              getDataURL: function () {
                return this.canvas.toDataURL()
              },
            }),
            n
          )
        })(),
        o = ((e = !1), 'canvas2d' === t.defaultRenderer && (e = r), e),
        i = function () {
          for (var e = {}, t = arguments.length, n = 0; n < t; n++) {
            var r = arguments[n]
            for (var o in r) e[o] = r[o]
          }
          return e
        },
        a = (function () {
          var e = (function () {
              function e() {
                this.cStore = {}
              }
              return (
                (e.prototype = {
                  on: function (e, t, n) {
                    var r = this.cStore
                    r[e] || (r[e] = []),
                      r[e].push(function (e) {
                        return t.call(n, e)
                      })
                  },
                  emit: function (e, t) {
                    var n = this.cStore
                    if (n[e]) for (var r = n[e].length, o = 0; o < r; o++) (0, n[e][o])(t)
                  },
                }),
                e
              )
            })(),
            r = function (e) {
              var t = e._renderer,
                n = e._coordinator,
                r = e._store
              n.on('renderpartial', t.renderPartial, t),
                n.on('renderall', t.renderAll, t),
                n.on('extremachange', function (t) {
                  e._config.onExtremaChange &&
                    e._config.onExtremaChange({
                      min: t.min,
                      max: t.max,
                      gradient: e._config.gradient || e._config.defaultGradient,
                    })
                }),
                r.setCoordinator(n)
            }
          function a() {
            var a = (this._config = i(t, arguments[0] || {}))
            if (((this._coordinator = new e()), a.plugin)) {
              var s = a.plugin
              if (!t.plugins[s]) throw new Error("Plugin '" + s + "' not found. Maybe it was not registered.")
              var u = t.plugins[s]
              ;(this._renderer = new u.renderer(a)), (this._store = new u.store(a))
            } else (this._renderer = new o(a)), (this._store = new n(a))
            r(this)
          }
          return (
            (a.prototype = {
              addData: function () {
                return this._store.addData.apply(this._store, arguments), this
              },
              removeData: function () {
                return this._store.removeData && this._store.removeData.apply(this._store, arguments), this
              },
              setData: function () {
                return this._store.setData.apply(this._store, arguments), this
              },
              setDataMax: function () {
                return this._store.setDataMax.apply(this._store, arguments), this
              },
              setDataMin: function () {
                return this._store.setDataMin.apply(this._store, arguments), this
              },
              configure: function (e) {
                return (
                  (this._config = i(this._config, e)),
                  this._renderer.updateConfig(this._config),
                  this._coordinator.emit('renderall', this._store._getInternalData()),
                  this
                )
              },
              repaint: function () {
                return this._coordinator.emit('renderall', this._store._getInternalData()), this
              },
              getData: function () {
                return this._store.getData()
              },
              getDataURL: function () {
                return this._renderer.getDataURL()
              },
              getValueAt: function (e) {
                return this._store.getValueAt
                  ? this._store.getValueAt(e)
                  : this._renderer.getValueAt
                  ? this._renderer.getValueAt(e)
                  : null
              },
            }),
            a
          )
        })()
      return {
        create: function (e) {
          return new a(e)
        },
        register: function (e, n) {
          t.plugins[e] = n
        },
      }
    }),
    /*
     * heatmap.js v2.0.5 | JavaScript Heatmap Library
     *
     * Copyright 2008-2016 Patrick Wied <heatmapjs@patrick-wied.at> - All rights reserved.
     * Dual licensed under MIT and Beerware license
     *
     * :: 2016-09-05 01:16
     */
    (o = u).exports
      ? (o.exports = a())
      : (i.h337 = a())
  var c = u.exports
  function l(e) {
    return 'object' == typeof e && null !== e
  }
  function f(e) {
    return void 0 === e
  }
  var d = { exports: {} },
    p = function (e, t) {
      return function () {
        for (var n = new Array(arguments.length), r = 0; r < n.length; r++) n[r] = arguments[r]
        return e.apply(t, n)
      }
    },
    h = p,
    m = Object.prototype.toString
  function v(e) {
    return '[object Array]' === m.call(e)
  }
  function g(e) {
    return void 0 === e
  }
  function y(e) {
    return null !== e && 'object' == typeof e
  }
  function b(e) {
    if ('[object Object]' !== m.call(e)) return !1
    var t = Object.getPrototypeOf(e)
    return null === t || t === Object.prototype
  }
  function w(e) {
    return '[object Function]' === m.call(e)
  }
  function x(e, t) {
    if (null != e)
      if (('object' != typeof e && (e = [e]), v(e))) for (var n = 0, r = e.length; n < r; n++) t.call(null, e[n], n, e)
      else for (var o in e) Object.prototype.hasOwnProperty.call(e, o) && t.call(null, e[o], o, e)
  }
  var _ = {
      isArray: v,
      isArrayBuffer: function (e) {
        return '[object ArrayBuffer]' === m.call(e)
      },
      isBuffer: function (e) {
        return (
          null !== e &&
          !g(e) &&
          null !== e.constructor &&
          !g(e.constructor) &&
          'function' == typeof e.constructor.isBuffer &&
          e.constructor.isBuffer(e)
        )
      },
      isFormData: function (e) {
        return 'undefined' != typeof FormData && e instanceof FormData
      },
      isArrayBufferView: function (e) {
        return 'undefined' != typeof ArrayBuffer && ArrayBuffer.isView
          ? ArrayBuffer.isView(e)
          : e && e.buffer && e.buffer instanceof ArrayBuffer
      },
      isString: function (e) {
        return 'string' == typeof e
      },
      isNumber: function (e) {
        return 'number' == typeof e
      },
      isObject: y,
      isPlainObject: b,
      isUndefined: g,
      isDate: function (e) {
        return '[object Date]' === m.call(e)
      },
      isFile: function (e) {
        return '[object File]' === m.call(e)
      },
      isBlob: function (e) {
        return '[object Blob]' === m.call(e)
      },
      isFunction: w,
      isStream: function (e) {
        return y(e) && w(e.pipe)
      },
      isURLSearchParams: function (e) {
        return 'undefined' != typeof URLSearchParams && e instanceof URLSearchParams
      },
      isStandardBrowserEnv: function () {
        return (
          ('undefined' == typeof navigator ||
            ('ReactNative' !== navigator.product &&
              'NativeScript' !== navigator.product &&
              'NS' !== navigator.product)) &&
          'undefined' != typeof window &&
          'undefined' != typeof document
        )
      },
      forEach: x,
      merge: function e() {
        var t = {}
        function n(n, r) {
          b(t[r]) && b(n) ? (t[r] = e(t[r], n)) : b(n) ? (t[r] = e({}, n)) : v(n) ? (t[r] = n.slice()) : (t[r] = n)
        }
        for (var r = 0, o = arguments.length; r < o; r++) x(arguments[r], n)
        return t
      },
      extend: function (e, t, n) {
        return (
          x(t, function (t, r) {
            e[r] = n && 'function' == typeof t ? h(t, n) : t
          }),
          e
        )
      },
      trim: function (e) {
        return e.trim ? e.trim() : e.replace(/^\s+|\s+$/g, '')
      },
      stripBOM: function (e) {
        return 65279 === e.charCodeAt(0) && (e = e.slice(1)), e
      },
    },
    O = _
  function E(e) {
    return encodeURIComponent(e)
      .replace(/%3A/gi, ':')
      .replace(/%24/g, '$')
      .replace(/%2C/gi, ',')
      .replace(/%20/g, '+')
      .replace(/%5B/gi, '[')
      .replace(/%5D/gi, ']')
  }
  var C = function (e, t, n) {
      if (!t) return e
      var r
      if (n) r = n(t)
      else if (O.isURLSearchParams(t)) r = t.toString()
      else {
        var o = []
        O.forEach(t, function (e, t) {
          null != e &&
            (O.isArray(e) ? (t += '[]') : (e = [e]),
            O.forEach(e, function (e) {
              O.isDate(e) ? (e = e.toISOString()) : O.isObject(e) && (e = JSON.stringify(e)), o.push(E(t) + '=' + E(e))
            }))
        }),
          (r = o.join('&'))
      }
      if (r) {
        var i = e.indexOf('#')
        ;-1 !== i && (e = e.slice(0, i)), (e += (-1 === e.indexOf('?') ? '?' : '&') + r)
      }
      return e
    },
    S = _
  function D() {
    this.handlers = []
  }
  ;(D.prototype.use = function (e, t, n) {
    return (
      this.handlers.push({
        fulfilled: e,
        rejected: t,
        synchronous: !!n && n.synchronous,
        runWhen: n ? n.runWhen : null,
      }),
      this.handlers.length - 1
    )
  }),
    (D.prototype.eject = function (e) {
      this.handlers[e] && (this.handlers[e] = null)
    }),
    (D.prototype.forEach = function (e) {
      S.forEach(this.handlers, function (t) {
        null !== t && e(t)
      })
    })
  var A = D,
    k = _,
    j = function (e, t, n, r, o) {
      return (
        (e.config = t),
        n && (e.code = n),
        (e.request = r),
        (e.response = o),
        (e.isAxiosError = !0),
        (e.toJSON = function () {
          return {
            message: this.message,
            name: this.name,
            description: this.description,
            number: this.number,
            fileName: this.fileName,
            lineNumber: this.lineNumber,
            columnNumber: this.columnNumber,
            stack: this.stack,
            config: this.config,
            code: this.code,
            status: this.response && this.response.status ? this.response.status : null,
          }
        }),
        e
      )
    },
    T = j,
    M = function (e, t, n, r, o) {
      var i = new Error(e)
      return T(i, t, n, r, o)
    },
    P = M,
    R = _,
    L = R.isStandardBrowserEnv()
      ? {
          write: function (e, t, n, r, o, i) {
            var a = []
            a.push(e + '=' + encodeURIComponent(t)),
              R.isNumber(n) && a.push('expires=' + new Date(n).toGMTString()),
              R.isString(r) && a.push('path=' + r),
              R.isString(o) && a.push('domain=' + o),
              !0 === i && a.push('secure'),
              (document.cookie = a.join('; '))
          },
          read: function (e) {
            var t = document.cookie.match(new RegExp('(^|;\\s*)(' + e + ')=([^;]*)'))
            return t ? decodeURIComponent(t[3]) : null
          },
          remove: function (e) {
            this.write(e, '', Date.now() - 864e5)
          },
        }
      : {
          write: function () {},
          read: function () {
            return null
          },
          remove: function () {},
        },
    N = function (e) {
      return /^([a-z][a-z\d\+\-\.]*:)?\/\//i.test(e)
    },
    B = function (e, t) {
      return t ? e.replace(/\/+$/, '') + '/' + t.replace(/^\/+/, '') : e
    },
    q = _,
    I = [
      'age',
      'authorization',
      'content-length',
      'content-type',
      'etag',
      'expires',
      'from',
      'host',
      'if-modified-since',
      'if-unmodified-since',
      'last-modified',
      'location',
      'max-forwards',
      'proxy-authorization',
      'referer',
      'retry-after',
      'user-agent',
    ],
    F = _,
    U = F.isStandardBrowserEnv()
      ? (function () {
          var e,
            t = /(msie|trident)/i.test(navigator.userAgent),
            n = document.createElement('a')
          function r(e) {
            var r = e
            return (
              t && (n.setAttribute('href', r), (r = n.href)),
              n.setAttribute('href', r),
              {
                href: n.href,
                protocol: n.protocol ? n.protocol.replace(/:$/, '') : '',
                host: n.host,
                search: n.search ? n.search.replace(/^\?/, '') : '',
                hash: n.hash ? n.hash.replace(/^#/, '') : '',
                hostname: n.hostname,
                port: n.port,
                pathname: '/' === n.pathname.charAt(0) ? n.pathname : '/' + n.pathname,
              }
            )
          }
          return (
            (e = r(window.location.href)),
            function (t) {
              var n = F.isString(t) ? r(t) : t
              return n.protocol === e.protocol && n.host === e.host
            }
          )
        })()
      : function () {
          return !0
        }
  function H(e) {
    this.message = e
  }
  ;(H.prototype.toString = function () {
    return 'Cancel' + (this.message ? ': ' + this.message : '')
  }),
    (H.prototype.__CANCEL__ = !0)
  var W = H,
    z = _,
    V = function (e, t, n) {
      var r = n.config.validateStatus
      n.status && r && !r(n.status)
        ? t(P('Request failed with status code ' + n.status, n.config, null, n.request, n))
        : e(n)
    },
    G = L,
    J = C,
    X = function (e, t) {
      return e && !N(t) ? B(e, t) : t
    },
    Y = function (e) {
      var t,
        n,
        r,
        o = {}
      return e
        ? (q.forEach(e.split('\n'), function (e) {
            if (((r = e.indexOf(':')), (t = q.trim(e.substr(0, r)).toLowerCase()), (n = q.trim(e.substr(r + 1))), t)) {
              if (o[t] && I.indexOf(t) >= 0) return
              o[t] = 'set-cookie' === t ? (o[t] ? o[t] : []).concat([n]) : o[t] ? o[t] + ', ' + n : n
            }
          }),
          o)
        : o
    },
    $ = U,
    K = M,
    Q = ue,
    Z = W,
    ee = function (e) {
      return new Promise(function (t, n) {
        var r,
          o = e.data,
          i = e.headers,
          a = e.responseType
        function s() {
          e.cancelToken && e.cancelToken.unsubscribe(r), e.signal && e.signal.removeEventListener('abort', r)
        }
        z.isFormData(o) && delete i['Content-Type']
        var u = new XMLHttpRequest()
        if (e.auth) {
          var c = e.auth.username || '',
            l = e.auth.password ? unescape(encodeURIComponent(e.auth.password)) : ''
          i.Authorization = 'Basic ' + btoa(c + ':' + l)
        }
        var f = X(e.baseURL, e.url)
        function d() {
          if (u) {
            var r = 'getAllResponseHeaders' in u ? Y(u.getAllResponseHeaders()) : null,
              o = {
                data: a && 'text' !== a && 'json' !== a ? u.response : u.responseText,
                status: u.status,
                statusText: u.statusText,
                headers: r,
                config: e,
                request: u,
              }
            V(
              function (e) {
                t(e), s()
              },
              function (e) {
                n(e), s()
              },
              o,
            ),
              (u = null)
          }
        }
        if (
          (u.open(e.method.toUpperCase(), J(f, e.params, e.paramsSerializer), !0),
          (u.timeout = e.timeout),
          'onloadend' in u
            ? (u.onloadend = d)
            : (u.onreadystatechange = function () {
                u &&
                  4 === u.readyState &&
                  (0 !== u.status || (u.responseURL && 0 === u.responseURL.indexOf('file:'))) &&
                  setTimeout(d)
              }),
          (u.onabort = function () {
            u && (n(K('Request aborted', e, 'ECONNABORTED', u)), (u = null))
          }),
          (u.onerror = function () {
            n(K('Network Error', e, null, u)), (u = null)
          }),
          (u.ontimeout = function () {
            var t = e.timeout ? 'timeout of ' + e.timeout + 'ms exceeded' : 'timeout exceeded',
              r = e.transitional || Q.transitional
            e.timeoutErrorMessage && (t = e.timeoutErrorMessage),
              n(K(t, e, r.clarifyTimeoutError ? 'ETIMEDOUT' : 'ECONNABORTED', u)),
              (u = null)
          }),
          z.isStandardBrowserEnv())
        ) {
          var p = (e.withCredentials || $(f)) && e.xsrfCookieName ? G.read(e.xsrfCookieName) : void 0
          p && (i[e.xsrfHeaderName] = p)
        }
        'setRequestHeader' in u &&
          z.forEach(i, function (e, t) {
            void 0 === o && 'content-type' === t.toLowerCase() ? delete i[t] : u.setRequestHeader(t, e)
          }),
          z.isUndefined(e.withCredentials) || (u.withCredentials = !!e.withCredentials),
          a && 'json' !== a && (u.responseType = e.responseType),
          'function' == typeof e.onDownloadProgress && u.addEventListener('progress', e.onDownloadProgress),
          'function' == typeof e.onUploadProgress &&
            u.upload &&
            u.upload.addEventListener('progress', e.onUploadProgress),
          (e.cancelToken || e.signal) &&
            ((r = function (e) {
              u && (n(!e || (e && e.type) ? new Z('canceled') : e), u.abort(), (u = null))
            }),
            e.cancelToken && e.cancelToken.subscribe(r),
            e.signal && (e.signal.aborted ? r() : e.signal.addEventListener('abort', r))),
          o || (o = null),
          u.send(o)
      })
    },
    te = _,
    ne = function (e, t) {
      k.forEach(e, function (n, r) {
        r !== t && r.toUpperCase() === t.toUpperCase() && ((e[t] = n), delete e[r])
      })
    },
    re = j,
    oe = { 'Content-Type': 'application/x-www-form-urlencoded' }
  function ie(e, t) {
    !te.isUndefined(e) && te.isUndefined(e['Content-Type']) && (e['Content-Type'] = t)
  }
  var ae,
    se = {
      transitional: { silentJSONParsing: !0, forcedJSONParsing: !0, clarifyTimeoutError: !1 },
      adapter:
        (('undefined' != typeof XMLHttpRequest ||
          ('undefined' != typeof process && '[object process]' === Object.prototype.toString.call(process))) &&
          (ae = ee),
        ae),
      transformRequest: [
        function (e, t) {
          return (
            ne(t, 'Accept'),
            ne(t, 'Content-Type'),
            te.isFormData(e) || te.isArrayBuffer(e) || te.isBuffer(e) || te.isStream(e) || te.isFile(e) || te.isBlob(e)
              ? e
              : te.isArrayBufferView(e)
              ? e.buffer
              : te.isURLSearchParams(e)
              ? (ie(t, 'application/x-www-form-urlencoded;charset=utf-8'), e.toString())
              : te.isObject(e) || (t && 'application/json' === t['Content-Type'])
              ? (ie(t, 'application/json'),
                (function (e, t, n) {
                  if (te.isString(e))
                    try {
                      return (t || JSON.parse)(e), te.trim(e)
                    } catch (e) {
                      if ('SyntaxError' !== e.name) throw e
                    }
                  return (n || JSON.stringify)(e)
                })(e))
              : e
          )
        },
      ],
      transformResponse: [
        function (e) {
          var t = this.transitional || se.transitional,
            n = t && t.silentJSONParsing,
            r = t && t.forcedJSONParsing,
            o = !n && 'json' === this.responseType
          if (o || (r && te.isString(e) && e.length))
            try {
              return JSON.parse(e)
            } catch (e) {
              if (o) {
                if ('SyntaxError' === e.name) throw re(e, this, 'E_JSON_PARSE')
                throw e
              }
            }
          return e
        },
      ],
      timeout: 0,
      xsrfCookieName: 'XSRF-TOKEN',
      xsrfHeaderName: 'X-XSRF-TOKEN',
      maxContentLength: -1,
      maxBodyLength: -1,
      validateStatus: function (e) {
        return e >= 200 && e < 300
      },
      headers: { common: { Accept: 'application/json, text/plain, */*' } },
    }
  te.forEach(['delete', 'get', 'head'], function (e) {
    se.headers[e] = {}
  }),
    te.forEach(['post', 'put', 'patch'], function (e) {
      se.headers[e] = te.merge(oe)
    })
  var ue = se,
    ce = _,
    le = ue,
    fe = function (e) {
      return !(!e || !e.__CANCEL__)
    },
    de = _,
    pe = function (e, t, n) {
      var r = this || le
      return (
        ce.forEach(n, function (n) {
          e = n.call(r, e, t)
        }),
        e
      )
    },
    he = fe,
    me = ue,
    ve = W
  function ge(e) {
    if ((e.cancelToken && e.cancelToken.throwIfRequested(), e.signal && e.signal.aborted)) throw new ve('canceled')
  }
  var ye = _,
    be = function (e, t) {
      t = t || {}
      var n = {}
      function r(e, t) {
        return ye.isPlainObject(e) && ye.isPlainObject(t)
          ? ye.merge(e, t)
          : ye.isPlainObject(t)
          ? ye.merge({}, t)
          : ye.isArray(t)
          ? t.slice()
          : t
      }
      function o(n) {
        return ye.isUndefined(t[n]) ? (ye.isUndefined(e[n]) ? void 0 : r(void 0, e[n])) : r(e[n], t[n])
      }
      function i(e) {
        if (!ye.isUndefined(t[e])) return r(void 0, t[e])
      }
      function a(n) {
        return ye.isUndefined(t[n]) ? (ye.isUndefined(e[n]) ? void 0 : r(void 0, e[n])) : r(void 0, t[n])
      }
      function s(n) {
        return n in t ? r(e[n], t[n]) : n in e ? r(void 0, e[n]) : void 0
      }
      var u = {
        url: i,
        method: i,
        data: i,
        baseURL: a,
        transformRequest: a,
        transformResponse: a,
        paramsSerializer: a,
        timeout: a,
        timeoutMessage: a,
        withCredentials: a,
        adapter: a,
        responseType: a,
        xsrfCookieName: a,
        xsrfHeaderName: a,
        onUploadProgress: a,
        onDownloadProgress: a,
        decompress: a,
        maxContentLength: a,
        maxBodyLength: a,
        transport: a,
        httpAgent: a,
        httpsAgent: a,
        cancelToken: a,
        socketPath: a,
        responseEncoding: a,
        validateStatus: s,
      }
      return (
        ye.forEach(Object.keys(e).concat(Object.keys(t)), function (e) {
          var t = u[e] || o,
            r = t(e)
          ;(ye.isUndefined(r) && t !== s) || (n[e] = r)
        }),
        n
      )
    },
    we = '0.24.0',
    xe = we,
    _e = {}
  ;['object', 'boolean', 'number', 'function', 'string', 'symbol'].forEach(function (e, t) {
    _e[e] = function (n) {
      return typeof n === e || 'a' + (t < 1 ? 'n ' : ' ') + e
    }
  })
  var Oe = {}
  _e.transitional = function (e, t, n) {
    function r(e, t) {
      return '[Axios v' + xe + "] Transitional option '" + e + "'" + t + (n ? '. ' + n : '')
    }
    return function (n, o, i) {
      if (!1 === e) throw new Error(r(o, ' has been removed' + (t ? ' in ' + t : '')))
      return (
        t &&
          !Oe[o] &&
          ((Oe[o] = !0),
          console.warn(r(o, ' has been deprecated since v' + t + ' and will be removed in the near future'))),
        !e || e(n, o, i)
      )
    }
  }
  var Ee = {
      assertOptions: function (e, t, n) {
        if ('object' != typeof e) throw new TypeError('options must be an object')
        for (var r = Object.keys(e), o = r.length; o-- > 0; ) {
          var i = r[o],
            a = t[i]
          if (a) {
            var s = e[i],
              u = void 0 === s || a(s, i, e)
            if (!0 !== u) throw new TypeError('option ' + i + ' must be ' + u)
          } else if (!0 !== n) throw Error('Unknown option ' + i)
        }
      },
      validators: _e,
    },
    Ce = _,
    Se = C,
    De = A,
    Ae = function (e) {
      return (
        ge(e),
        (e.headers = e.headers || {}),
        (e.data = pe.call(e, e.data, e.headers, e.transformRequest)),
        (e.headers = de.merge(e.headers.common || {}, e.headers[e.method] || {}, e.headers)),
        de.forEach(['delete', 'get', 'head', 'post', 'put', 'patch', 'common'], function (t) {
          delete e.headers[t]
        }),
        (e.adapter || me.adapter)(e).then(
          function (t) {
            return ge(e), (t.data = pe.call(e, t.data, t.headers, e.transformResponse)), t
          },
          function (t) {
            return (
              he(t) ||
                (ge(e),
                t &&
                  t.response &&
                  (t.response.data = pe.call(e, t.response.data, t.response.headers, e.transformResponse))),
              Promise.reject(t)
            )
          },
        )
      )
    },
    ke = be,
    je = Ee,
    Te = je.validators
  function Me(e) {
    ;(this.defaults = e), (this.interceptors = { request: new De(), response: new De() })
  }
  ;(Me.prototype.request = function (e) {
    'string' == typeof e ? ((e = arguments[1] || {}).url = arguments[0]) : (e = e || {}),
      (e = ke(this.defaults, e)).method
        ? (e.method = e.method.toLowerCase())
        : this.defaults.method
        ? (e.method = this.defaults.method.toLowerCase())
        : (e.method = 'get')
    var t = e.transitional
    void 0 !== t &&
      je.assertOptions(
        t,
        {
          silentJSONParsing: Te.transitional(Te.boolean),
          forcedJSONParsing: Te.transitional(Te.boolean),
          clarifyTimeoutError: Te.transitional(Te.boolean),
        },
        !1,
      )
    var n = [],
      r = !0
    this.interceptors.request.forEach(function (t) {
      ;('function' == typeof t.runWhen && !1 === t.runWhen(e)) ||
        ((r = r && t.synchronous), n.unshift(t.fulfilled, t.rejected))
    })
    var o,
      i = []
    if (
      (this.interceptors.response.forEach(function (e) {
        i.push(e.fulfilled, e.rejected)
      }),
      !r)
    ) {
      var a = [Ae, void 0]
      for (Array.prototype.unshift.apply(a, n), a = a.concat(i), o = Promise.resolve(e); a.length; )
        o = o.then(a.shift(), a.shift())
      return o
    }
    for (var s = e; n.length; ) {
      var u = n.shift(),
        c = n.shift()
      try {
        s = u(s)
      } catch (e) {
        c(e)
        break
      }
    }
    try {
      o = Ae(s)
    } catch (e) {
      return Promise.reject(e)
    }
    for (; i.length; ) o = o.then(i.shift(), i.shift())
    return o
  }),
    (Me.prototype.getUri = function (e) {
      return (e = ke(this.defaults, e)), Se(e.url, e.params, e.paramsSerializer).replace(/^\?/, '')
    }),
    Ce.forEach(['delete', 'get', 'head', 'options'], function (e) {
      Me.prototype[e] = function (t, n) {
        return this.request(ke(n || {}, { method: e, url: t, data: (n || {}).data }))
      }
    }),
    Ce.forEach(['post', 'put', 'patch'], function (e) {
      Me.prototype[e] = function (t, n, r) {
        return this.request(ke(r || {}, { method: e, url: t, data: n }))
      }
    })
  var Pe = Me,
    Re = W
  function Le(e) {
    if ('function' != typeof e) throw new TypeError('executor must be a function.')
    var t
    this.promise = new Promise(function (e) {
      t = e
    })
    var n = this
    this.promise.then(function (e) {
      if (n._listeners) {
        var t,
          r = n._listeners.length
        for (t = 0; t < r; t++) n._listeners[t](e)
        n._listeners = null
      }
    }),
      (this.promise.then = function (e) {
        var t,
          r = new Promise(function (e) {
            n.subscribe(e), (t = e)
          }).then(e)
        return (
          (r.cancel = function () {
            n.unsubscribe(t)
          }),
          r
        )
      }),
      e(function (e) {
        n.reason || ((n.reason = new Re(e)), t(n.reason))
      })
  }
  ;(Le.prototype.throwIfRequested = function () {
    if (this.reason) throw this.reason
  }),
    (Le.prototype.subscribe = function (e) {
      this.reason ? e(this.reason) : this._listeners ? this._listeners.push(e) : (this._listeners = [e])
    }),
    (Le.prototype.unsubscribe = function (e) {
      if (this._listeners) {
        var t = this._listeners.indexOf(e)
        ;-1 !== t && this._listeners.splice(t, 1)
      }
    }),
    (Le.source = function () {
      var e
      return {
        token: new Le(function (t) {
          e = t
        }),
        cancel: e,
      }
    })
  var Ne = Le,
    Be = _,
    qe = p,
    Ie = Pe,
    Fe = be
  var Ue = (function e(t) {
    var n = new Ie(t),
      r = qe(Ie.prototype.request, n)
    return (
      Be.extend(r, Ie.prototype, n),
      Be.extend(r, n),
      (r.create = function (n) {
        return e(Fe(t, n))
      }),
      r
    )
  })(ue)
  ;(Ue.Axios = Ie),
    (Ue.Cancel = W),
    (Ue.CancelToken = Ne),
    (Ue.isCancel = fe),
    (Ue.VERSION = we),
    (Ue.all = function (e) {
      return Promise.all(e)
    }),
    (Ue.spread = function (e) {
      return function (t) {
        return e.apply(null, t)
      }
    }),
    (Ue.isAxiosError = function (e) {
      return 'object' == typeof e && !0 === e.isAxiosError
    }),
    (d.exports = Ue),
    (d.exports.default = Ue)
  var He = d.exports,
    We = 'top',
    ze = 'bottom',
    Ve = 'right',
    Ge = 'left',
    Je = 'auto',
    Xe = [We, ze, Ve, Ge],
    Ye = 'start',
    $e = 'end',
    Ke = 'viewport',
    Qe = 'popper',
    Ze = Xe.reduce(function (e, t) {
      return e.concat([t + '-' + Ye, t + '-' + $e])
    }, []),
    et = [].concat(Xe, [Je]).reduce(function (e, t) {
      return e.concat([t, t + '-' + Ye, t + '-' + $e])
    }, []),
    tt = ['beforeRead', 'read', 'afterRead', 'beforeMain', 'main', 'afterMain', 'beforeWrite', 'write', 'afterWrite']
  function nt(e) {
    return e ? (e.nodeName || '').toLowerCase() : null
  }
  function rt(e) {
    if (null == e) return window
    if ('[object Window]' !== e.toString()) {
      var t = e.ownerDocument
      return (t && t.defaultView) || window
    }
    return e
  }
  function ot(e) {
    return e instanceof rt(e).Element || e instanceof Element
  }
  function it(e) {
    return e instanceof rt(e).HTMLElement || e instanceof HTMLElement
  }
  function at(e) {
    return 'undefined' != typeof ShadowRoot && (e instanceof rt(e).ShadowRoot || e instanceof ShadowRoot)
  }
  var st = {
    name: 'applyStyles',
    enabled: !0,
    phase: 'write',
    fn: function (e) {
      var t = e.state
      Object.keys(t.elements).forEach(function (e) {
        var n = t.styles[e] || {},
          r = t.attributes[e] || {},
          o = t.elements[e]
        it(o) &&
          nt(o) &&
          (Object.assign(o.style, n),
          Object.keys(r).forEach(function (e) {
            var t = r[e]
            !1 === t ? o.removeAttribute(e) : o.setAttribute(e, !0 === t ? '' : t)
          }))
      })
    },
    effect: function (e) {
      var t = e.state,
        n = {
          popper: { position: t.options.strategy, left: '0', top: '0', margin: '0' },
          arrow: { position: 'absolute' },
          reference: {},
        }
      return (
        Object.assign(t.elements.popper.style, n.popper),
        (t.styles = n),
        t.elements.arrow && Object.assign(t.elements.arrow.style, n.arrow),
        function () {
          Object.keys(t.elements).forEach(function (e) {
            var r = t.elements[e],
              o = t.attributes[e] || {},
              i = Object.keys(t.styles.hasOwnProperty(e) ? t.styles[e] : n[e]).reduce(function (e, t) {
                return (e[t] = ''), e
              }, {})
            it(r) &&
              nt(r) &&
              (Object.assign(r.style, i),
              Object.keys(o).forEach(function (e) {
                r.removeAttribute(e)
              }))
          })
        }
      )
    },
    requires: ['computeStyles'],
  }
  function ut(e) {
    return e.split('-')[0]
  }
  var ct = Math.round
  function lt(e, t) {
    void 0 === t && (t = !1)
    var n = e.getBoundingClientRect(),
      r = 1,
      o = 1
    if (it(e) && t) {
      var i = e.offsetHeight,
        a = e.offsetWidth
      a > 0 && (r = n.width / a || 1), i > 0 && (o = n.height / i || 1)
    }
    return {
      width: ct(n.width / r),
      height: ct(n.height / o),
      top: ct(n.top / o),
      right: ct(n.right / r),
      bottom: ct(n.bottom / o),
      left: ct(n.left / r),
      x: ct(n.left / r),
      y: ct(n.top / o),
    }
  }
  function ft(e) {
    var t = lt(e),
      n = e.offsetWidth,
      r = e.offsetHeight
    return (
      Math.abs(t.width - n) <= 1 && (n = t.width),
      Math.abs(t.height - r) <= 1 && (r = t.height),
      { x: e.offsetLeft, y: e.offsetTop, width: n, height: r }
    )
  }
  function dt(e, t) {
    var n = t.getRootNode && t.getRootNode()
    if (e.contains(t)) return !0
    if (n && at(n)) {
      var r = t
      do {
        if (r && e.isSameNode(r)) return !0
        r = r.parentNode || r.host
      } while (r)
    }
    return !1
  }
  function pt(e) {
    return rt(e).getComputedStyle(e)
  }
  function ht(e) {
    return ['table', 'td', 'th'].indexOf(nt(e)) >= 0
  }
  function mt(e) {
    return ((ot(e) ? e.ownerDocument : e.document) || window.document).documentElement
  }
  function vt(e) {
    return 'html' === nt(e) ? e : e.assignedSlot || e.parentNode || (at(e) ? e.host : null) || mt(e)
  }
  function gt(e) {
    return it(e) && 'fixed' !== pt(e).position ? e.offsetParent : null
  }
  function yt(e) {
    for (var t = rt(e), n = gt(e); n && ht(n) && 'static' === pt(n).position; ) n = gt(n)
    return n && ('html' === nt(n) || ('body' === nt(n) && 'static' === pt(n).position))
      ? t
      : n ||
          (function (e) {
            var t = -1 !== navigator.userAgent.toLowerCase().indexOf('firefox')
            if (-1 !== navigator.userAgent.indexOf('Trident') && it(e) && 'fixed' === pt(e).position) return null
            for (var n = vt(e); it(n) && ['html', 'body'].indexOf(nt(n)) < 0; ) {
              var r = pt(n)
              if (
                'none' !== r.transform ||
                'none' !== r.perspective ||
                'paint' === r.contain ||
                -1 !== ['transform', 'perspective'].indexOf(r.willChange) ||
                (t && 'filter' === r.willChange) ||
                (t && r.filter && 'none' !== r.filter)
              )
                return n
              n = n.parentNode
            }
            return null
          })(e) ||
          t
  }
  function bt(e) {
    return ['top', 'bottom'].indexOf(e) >= 0 ? 'x' : 'y'
  }
  var wt = Math.max,
    xt = Math.min,
    _t = Math.round
  function Ot(e, t, n) {
    return wt(e, xt(t, n))
  }
  function Et(e) {
    return Object.assign({}, { top: 0, right: 0, bottom: 0, left: 0 }, e)
  }
  function Ct(e, t) {
    return t.reduce(function (t, n) {
      return (t[n] = e), t
    }, {})
  }
  var St = {
    name: 'arrow',
    enabled: !0,
    phase: 'main',
    fn: function (e) {
      var t,
        n = e.state,
        r = e.name,
        o = e.options,
        i = n.elements.arrow,
        a = n.modifiersData.popperOffsets,
        s = ut(n.placement),
        u = bt(s),
        c = [Ge, Ve].indexOf(s) >= 0 ? 'height' : 'width'
      if (i && a) {
        var l = (function (e, t) {
            return Et(
              'number' !=
                typeof (e = 'function' == typeof e ? e(Object.assign({}, t.rects, { placement: t.placement })) : e)
                ? e
                : Ct(e, Xe),
            )
          })(o.padding, n),
          f = ft(i),
          d = 'y' === u ? We : Ge,
          p = 'y' === u ? ze : Ve,
          h = n.rects.reference[c] + n.rects.reference[u] - a[u] - n.rects.popper[c],
          m = a[u] - n.rects.reference[u],
          v = yt(i),
          g = v ? ('y' === u ? v.clientHeight || 0 : v.clientWidth || 0) : 0,
          y = h / 2 - m / 2,
          b = l[d],
          w = g - f[c] - l[p],
          x = g / 2 - f[c] / 2 + y,
          _ = Ot(b, x, w),
          O = u
        n.modifiersData[r] = (((t = {})[O] = _), (t.centerOffset = _ - x), t)
      }
    },
    effect: function (e) {
      var t = e.state,
        n = e.options.element,
        r = void 0 === n ? '[data-popper-arrow]' : n
      null != r &&
        ('string' != typeof r || (r = t.elements.popper.querySelector(r))) &&
        (it(r) ||
          console.error(
            [
              'Popper: "arrow" element must be an HTMLElement (not an SVGElement).',
              'To use an SVG arrow, wrap it in an HTMLElement that will be used as',
              'the arrow.',
            ].join(' '),
          ),
        dt(t.elements.popper, r)
          ? (t.elements.arrow = r)
          : console.error(
              ['Popper: "arrow" modifier\'s `element` must be a child of the popper', 'element.'].join(' '),
            ))
    },
    requires: ['popperOffsets'],
    requiresIfExists: ['preventOverflow'],
  }
  function Dt(e) {
    return e.split('-')[1]
  }
  var At = { top: 'auto', right: 'auto', bottom: 'auto', left: 'auto' }
  function kt(e) {
    var t,
      n = e.popper,
      r = e.popperRect,
      o = e.placement,
      i = e.variation,
      a = e.offsets,
      s = e.position,
      u = e.gpuAcceleration,
      c = e.adaptive,
      l = e.roundOffsets,
      f =
        !0 === l
          ? (function (e) {
              var t = e.x,
                n = e.y,
                r = window.devicePixelRatio || 1
              return { x: _t(_t(t * r) / r) || 0, y: _t(_t(n * r) / r) || 0 }
            })(a)
          : 'function' == typeof l
          ? l(a)
          : a,
      d = f.x,
      p = void 0 === d ? 0 : d,
      h = f.y,
      m = void 0 === h ? 0 : h,
      v = a.hasOwnProperty('x'),
      g = a.hasOwnProperty('y'),
      y = Ge,
      b = We,
      w = window
    if (c) {
      var x = yt(n),
        _ = 'clientHeight',
        O = 'clientWidth'
      x === rt(n) &&
        'static' !== pt((x = mt(n))).position &&
        'absolute' === s &&
        ((_ = 'scrollHeight'), (O = 'scrollWidth')),
        (x = x),
        (o !== We && ((o !== Ge && o !== Ve) || i !== $e)) || ((b = ze), (m -= x[_] - r.height), (m *= u ? 1 : -1)),
        (o !== Ge && ((o !== We && o !== ze) || i !== $e)) || ((y = Ve), (p -= x[O] - r.width), (p *= u ? 1 : -1))
    }
    var E,
      C = Object.assign({ position: s }, c && At)
    return u
      ? Object.assign(
          {},
          C,
          (((E = {})[b] = g ? '0' : ''),
          (E[y] = v ? '0' : ''),
          (E.transform =
            (w.devicePixelRatio || 1) <= 1
              ? 'translate(' + p + 'px, ' + m + 'px)'
              : 'translate3d(' + p + 'px, ' + m + 'px, 0)'),
          E),
        )
      : Object.assign({}, C, (((t = {})[b] = g ? m + 'px' : ''), (t[y] = v ? p + 'px' : ''), (t.transform = ''), t))
  }
  var jt = { passive: !0 }
  var Tt = { left: 'right', right: 'left', bottom: 'top', top: 'bottom' }
  function Mt(e) {
    return e.replace(/left|right|bottom|top/g, function (e) {
      return Tt[e]
    })
  }
  var Pt = { start: 'end', end: 'start' }
  function Rt(e) {
    return e.replace(/start|end/g, function (e) {
      return Pt[e]
    })
  }
  function Lt(e) {
    var t = rt(e)
    return { scrollLeft: t.pageXOffset, scrollTop: t.pageYOffset }
  }
  function Nt(e) {
    return lt(mt(e)).left + Lt(e).scrollLeft
  }
  function Bt(e) {
    var t = pt(e),
      n = t.overflow,
      r = t.overflowX,
      o = t.overflowY
    return /auto|scroll|overlay|hidden/.test(n + o + r)
  }
  function qt(e) {
    return ['html', 'body', '#document'].indexOf(nt(e)) >= 0 ? e.ownerDocument.body : it(e) && Bt(e) ? e : qt(vt(e))
  }
  function It(e, t) {
    var n
    void 0 === t && (t = [])
    var r = qt(e),
      o = r === (null == (n = e.ownerDocument) ? void 0 : n.body),
      i = rt(r),
      a = o ? [i].concat(i.visualViewport || [], Bt(r) ? r : []) : r,
      s = t.concat(a)
    return o ? s : s.concat(It(vt(a)))
  }
  function Ft(e) {
    return Object.assign({}, e, { left: e.x, top: e.y, right: e.x + e.width, bottom: e.y + e.height })
  }
  function Ut(e, t) {
    return t === Ke
      ? Ft(
          (function (e) {
            var t = rt(e),
              n = mt(e),
              r = t.visualViewport,
              o = n.clientWidth,
              i = n.clientHeight,
              a = 0,
              s = 0
            return (
              r &&
                ((o = r.width),
                (i = r.height),
                /^((?!chrome|android).)*safari/i.test(navigator.userAgent) || ((a = r.offsetLeft), (s = r.offsetTop))),
              { width: o, height: i, x: a + Nt(e), y: s }
            )
          })(e),
        )
      : it(t)
      ? (function (e) {
          var t = lt(e)
          return (
            (t.top = t.top + e.clientTop),
            (t.left = t.left + e.clientLeft),
            (t.bottom = t.top + e.clientHeight),
            (t.right = t.left + e.clientWidth),
            (t.width = e.clientWidth),
            (t.height = e.clientHeight),
            (t.x = t.left),
            (t.y = t.top),
            t
          )
        })(t)
      : Ft(
          (function (e) {
            var t,
              n = mt(e),
              r = Lt(e),
              o = null == (t = e.ownerDocument) ? void 0 : t.body,
              i = wt(n.scrollWidth, n.clientWidth, o ? o.scrollWidth : 0, o ? o.clientWidth : 0),
              a = wt(n.scrollHeight, n.clientHeight, o ? o.scrollHeight : 0, o ? o.clientHeight : 0),
              s = -r.scrollLeft + Nt(e),
              u = -r.scrollTop
            return (
              'rtl' === pt(o || n).direction && (s += wt(n.clientWidth, o ? o.clientWidth : 0) - i),
              { width: i, height: a, x: s, y: u }
            )
          })(mt(e)),
        )
  }
  function Ht(e, t, n) {
    var r =
        'clippingParents' === t
          ? (function (e) {
              var t = It(vt(e)),
                n = ['absolute', 'fixed'].indexOf(pt(e).position) >= 0 && it(e) ? yt(e) : e
              return ot(n)
                ? t.filter(function (e) {
                    return ot(e) && dt(e, n) && 'body' !== nt(e)
                  })
                : []
            })(e)
          : [].concat(t),
      o = [].concat(r, [n]),
      i = o[0],
      a = o.reduce(function (t, n) {
        var r = Ut(e, n)
        return (
          (t.top = wt(r.top, t.top)),
          (t.right = xt(r.right, t.right)),
          (t.bottom = xt(r.bottom, t.bottom)),
          (t.left = wt(r.left, t.left)),
          t
        )
      }, Ut(e, i))
    return (a.width = a.right - a.left), (a.height = a.bottom - a.top), (a.x = a.left), (a.y = a.top), a
  }
  function Wt(e) {
    var t,
      n = e.reference,
      r = e.element,
      o = e.placement,
      i = o ? ut(o) : null,
      a = o ? Dt(o) : null,
      s = n.x + n.width / 2 - r.width / 2,
      u = n.y + n.height / 2 - r.height / 2
    switch (i) {
      case We:
        t = { x: s, y: n.y - r.height }
        break
      case ze:
        t = { x: s, y: n.y + n.height }
        break
      case Ve:
        t = { x: n.x + n.width, y: u }
        break
      case Ge:
        t = { x: n.x - r.width, y: u }
        break
      default:
        t = { x: n.x, y: n.y }
    }
    var c = i ? bt(i) : null
    if (null != c) {
      var l = 'y' === c ? 'height' : 'width'
      switch (a) {
        case Ye:
          t[c] = t[c] - (n[l] / 2 - r[l] / 2)
          break
        case $e:
          t[c] = t[c] + (n[l] / 2 - r[l] / 2)
      }
    }
    return t
  }
  function zt(e, t) {
    void 0 === t && (t = {})
    var n = t,
      r = n.placement,
      o = void 0 === r ? e.placement : r,
      i = n.boundary,
      a = void 0 === i ? 'clippingParents' : i,
      s = n.rootBoundary,
      u = void 0 === s ? Ke : s,
      c = n.elementContext,
      l = void 0 === c ? Qe : c,
      f = n.altBoundary,
      d = void 0 !== f && f,
      p = n.padding,
      h = void 0 === p ? 0 : p,
      m = Et('number' != typeof h ? h : Ct(h, Xe)),
      v = l === Qe ? 'reference' : Qe,
      g = e.rects.popper,
      y = e.elements[d ? v : l],
      b = Ht(ot(y) ? y : y.contextElement || mt(e.elements.popper), a, u),
      w = lt(e.elements.reference),
      x = Wt({ reference: w, element: g, strategy: 'absolute', placement: o }),
      _ = Ft(Object.assign({}, g, x)),
      O = l === Qe ? _ : w,
      E = {
        top: b.top - O.top + m.top,
        bottom: O.bottom - b.bottom + m.bottom,
        left: b.left - O.left + m.left,
        right: O.right - b.right + m.right,
      },
      C = e.modifiersData.offset
    if (l === Qe && C) {
      var S = C[o]
      Object.keys(E).forEach(function (e) {
        var t = [Ve, ze].indexOf(e) >= 0 ? 1 : -1,
          n = [We, ze].indexOf(e) >= 0 ? 'y' : 'x'
        E[e] += S[n] * t
      })
    }
    return E
  }
  function Vt(e, t) {
    void 0 === t && (t = {})
    var n = t,
      r = n.placement,
      o = n.boundary,
      i = n.rootBoundary,
      a = n.padding,
      s = n.flipVariations,
      u = n.allowedAutoPlacements,
      c = void 0 === u ? et : u,
      l = Dt(r),
      f = l
        ? s
          ? Ze
          : Ze.filter(function (e) {
              return Dt(e) === l
            })
        : Xe,
      d = f.filter(function (e) {
        return c.indexOf(e) >= 0
      })
    0 === d.length &&
      ((d = f),
      console.error(
        [
          'Popper: The `allowedAutoPlacements` option did not allow any',
          'placements. Ensure the `placement` option matches the variation',
          'of the allowed placements.',
          'For example, "auto" cannot be used to allow "bottom-start".',
          'Use "auto-start" instead.',
        ].join(' '),
      ))
    var p = d.reduce(function (t, n) {
      return (t[n] = zt(e, { placement: n, boundary: o, rootBoundary: i, padding: a })[ut(n)]), t
    }, {})
    return Object.keys(p).sort(function (e, t) {
      return p[e] - p[t]
    })
  }
  var Gt = {
    name: 'flip',
    enabled: !0,
    phase: 'main',
    fn: function (e) {
      var t = e.state,
        n = e.options,
        r = e.name
      if (!t.modifiersData[r]._skip) {
        for (
          var o = n.mainAxis,
            i = void 0 === o || o,
            a = n.altAxis,
            s = void 0 === a || a,
            u = n.fallbackPlacements,
            c = n.padding,
            l = n.boundary,
            f = n.rootBoundary,
            d = n.altBoundary,
            p = n.flipVariations,
            h = void 0 === p || p,
            m = n.allowedAutoPlacements,
            v = t.options.placement,
            g = ut(v),
            y =
              u ||
              (g === v || !h
                ? [Mt(v)]
                : (function (e) {
                    if (ut(e) === Je) return []
                    var t = Mt(e)
                    return [Rt(e), t, Rt(t)]
                  })(v)),
            b = [v].concat(y).reduce(function (e, n) {
              return e.concat(
                ut(n) === Je
                  ? Vt(t, {
                      placement: n,
                      boundary: l,
                      rootBoundary: f,
                      padding: c,
                      flipVariations: h,
                      allowedAutoPlacements: m,
                    })
                  : n,
              )
            }, []),
            w = t.rects.reference,
            x = t.rects.popper,
            _ = new Map(),
            O = !0,
            E = b[0],
            C = 0;
          C < b.length;
          C++
        ) {
          var S = b[C],
            D = ut(S),
            A = Dt(S) === Ye,
            k = [We, ze].indexOf(D) >= 0,
            j = k ? 'width' : 'height',
            T = zt(t, { placement: S, boundary: l, rootBoundary: f, altBoundary: d, padding: c }),
            M = k ? (A ? Ve : Ge) : A ? ze : We
          w[j] > x[j] && (M = Mt(M))
          var P = Mt(M),
            R = []
          if (
            (i && R.push(T[D] <= 0),
            s && R.push(T[M] <= 0, T[P] <= 0),
            R.every(function (e) {
              return e
            }))
          ) {
            ;(E = S), (O = !1)
            break
          }
          _.set(S, R)
        }
        if (O)
          for (
            var L = function (e) {
                var t = b.find(function (t) {
                  var n = _.get(t)
                  if (n)
                    return n.slice(0, e).every(function (e) {
                      return e
                    })
                })
                if (t) return (E = t), 'break'
              },
              N = h ? 3 : 1;
            N > 0;
            N--
          ) {
            if ('break' === L(N)) break
          }
        t.placement !== E && ((t.modifiersData[r]._skip = !0), (t.placement = E), (t.reset = !0))
      }
    },
    requiresIfExists: ['offset'],
    data: { _skip: !1 },
  }
  function Jt(e, t, n) {
    return (
      void 0 === n && (n = { x: 0, y: 0 }),
      {
        top: e.top - t.height - n.y,
        right: e.right - t.width + n.x,
        bottom: e.bottom - t.height + n.y,
        left: e.left - t.width - n.x,
      }
    )
  }
  function Xt(e) {
    return [We, Ve, ze, Ge].some(function (t) {
      return e[t] >= 0
    })
  }
  var Yt = {
    name: 'offset',
    enabled: !0,
    phase: 'main',
    requires: ['popperOffsets'],
    fn: function (e) {
      var t = e.state,
        n = e.options,
        r = e.name,
        o = n.offset,
        i = void 0 === o ? [0, 0] : o,
        a = et.reduce(function (e, n) {
          return (
            (e[n] = (function (e, t, n) {
              var r = ut(e),
                o = [Ge, We].indexOf(r) >= 0 ? -1 : 1,
                i = 'function' == typeof n ? n(Object.assign({}, t, { placement: e })) : n,
                a = i[0],
                s = i[1]
              return (a = a || 0), (s = (s || 0) * o), [Ge, Ve].indexOf(r) >= 0 ? { x: s, y: a } : { x: a, y: s }
            })(n, t.rects, i)),
            e
          )
        }, {}),
        s = a[t.placement],
        u = s.x,
        c = s.y
      null != t.modifiersData.popperOffsets &&
        ((t.modifiersData.popperOffsets.x += u), (t.modifiersData.popperOffsets.y += c)),
        (t.modifiersData[r] = a)
    },
  }
  var $t = {
    name: 'preventOverflow',
    enabled: !0,
    phase: 'main',
    fn: function (e) {
      var t = e.state,
        n = e.options,
        r = e.name,
        o = n.mainAxis,
        i = void 0 === o || o,
        a = n.altAxis,
        s = void 0 !== a && a,
        u = n.boundary,
        c = n.rootBoundary,
        l = n.altBoundary,
        f = n.padding,
        d = n.tether,
        p = void 0 === d || d,
        h = n.tetherOffset,
        m = void 0 === h ? 0 : h,
        v = zt(t, { boundary: u, rootBoundary: c, padding: f, altBoundary: l }),
        g = ut(t.placement),
        y = Dt(t.placement),
        b = !y,
        w = bt(g),
        x = 'x' === w ? 'y' : 'x',
        _ = t.modifiersData.popperOffsets,
        O = t.rects.reference,
        E = t.rects.popper,
        C = 'function' == typeof m ? m(Object.assign({}, t.rects, { placement: t.placement })) : m,
        S = { x: 0, y: 0 }
      if (_) {
        if (i || s) {
          var D = 'y' === w ? We : Ge,
            A = 'y' === w ? ze : Ve,
            k = 'y' === w ? 'height' : 'width',
            j = _[w],
            T = _[w] + v[D],
            M = _[w] - v[A],
            P = p ? -E[k] / 2 : 0,
            R = y === Ye ? O[k] : E[k],
            L = y === Ye ? -E[k] : -O[k],
            N = t.elements.arrow,
            B = p && N ? ft(N) : { width: 0, height: 0 },
            q = t.modifiersData['arrow#persistent']
              ? t.modifiersData['arrow#persistent'].padding
              : { top: 0, right: 0, bottom: 0, left: 0 },
            I = q[D],
            F = q[A],
            U = Ot(0, O[k], B[k]),
            H = b ? O[k] / 2 - P - U - I - C : R - U - I - C,
            W = b ? -O[k] / 2 + P + U + F + C : L + U + F + C,
            z = t.elements.arrow && yt(t.elements.arrow),
            V = z ? ('y' === w ? z.clientTop || 0 : z.clientLeft || 0) : 0,
            G = t.modifiersData.offset ? t.modifiersData.offset[t.placement][w] : 0,
            J = _[w] + H - G - V,
            X = _[w] + W - G
          if (i) {
            var Y = Ot(p ? xt(T, J) : T, j, p ? wt(M, X) : M)
            ;(_[w] = Y), (S[w] = Y - j)
          }
          if (s) {
            var $ = 'x' === w ? We : Ge,
              K = 'x' === w ? ze : Ve,
              Q = _[x],
              Z = Q + v[$],
              ee = Q - v[K],
              te = Ot(p ? xt(Z, J) : Z, Q, p ? wt(ee, X) : ee)
            ;(_[x] = te), (S[x] = te - Q)
          }
        }
        t.modifiersData[r] = S
      }
    },
    requiresIfExists: ['offset'],
  }
  function Kt(e, t, n) {
    void 0 === n && (n = !1)
    var r,
      o,
      i = it(t),
      a =
        it(t) &&
        (function (e) {
          var t = e.getBoundingClientRect(),
            n = t.width / e.offsetWidth || 1,
            r = t.height / e.offsetHeight || 1
          return 1 !== n || 1 !== r
        })(t),
      s = mt(t),
      u = lt(e, a),
      c = { scrollLeft: 0, scrollTop: 0 },
      l = { x: 0, y: 0 }
    return (
      (i || (!i && !n)) &&
        (('body' !== nt(t) || Bt(s)) &&
          (c = (r = t) !== rt(r) && it(r) ? { scrollLeft: (o = r).scrollLeft, scrollTop: o.scrollTop } : Lt(r)),
        it(t) ? (((l = lt(t, !0)).x += t.clientLeft), (l.y += t.clientTop)) : s && (l.x = Nt(s))),
      { x: u.left + c.scrollLeft - l.x, y: u.top + c.scrollTop - l.y, width: u.width, height: u.height }
    )
  }
  function Qt(e) {
    var t = new Map(),
      n = new Set(),
      r = []
    function o(e) {
      n.add(e.name),
        [].concat(e.requires || [], e.requiresIfExists || []).forEach(function (e) {
          if (!n.has(e)) {
            var r = t.get(e)
            r && o(r)
          }
        }),
        r.push(e)
    }
    return (
      e.forEach(function (e) {
        t.set(e.name, e)
      }),
      e.forEach(function (e) {
        n.has(e.name) || o(e)
      }),
      r
    )
  }
  function Zt(e) {
    for (var t = arguments.length, n = new Array(t > 1 ? t - 1 : 0), r = 1; r < t; r++) n[r - 1] = arguments[r]
    return [].concat(n).reduce(function (e, t) {
      return e.replace(/%s/, t)
    }, e)
  }
  var en = 'Popper: modifier "%s" provided an invalid %s property, expected %s but got %s',
    tn = ['name', 'enabled', 'phase', 'fn', 'effect', 'requires', 'options']
  var nn =
      'Popper: Invalid reference or popper argument provided. They must be either a DOM element or virtual element.',
    rn = { placement: 'bottom', modifiers: [], strategy: 'absolute' }
  function on() {
    for (var e = arguments.length, t = new Array(e), n = 0; n < e; n++) t[n] = arguments[n]
    return !t.some(function (e) {
      return !(e && 'function' == typeof e.getBoundingClientRect)
    })
  }
  function an(e) {
    void 0 === e && (e = {})
    var t = e,
      n = t.defaultModifiers,
      r = void 0 === n ? [] : n,
      o = t.defaultOptions,
      i = void 0 === o ? rn : o
    return function (e, t, n) {
      void 0 === n && (n = i)
      var o,
        a,
        s = {
          placement: 'bottom',
          orderedModifiers: [],
          options: Object.assign({}, rn, i),
          modifiersData: {},
          elements: { reference: e, popper: t },
          attributes: {},
          styles: {},
        },
        u = [],
        c = !1,
        l = {
          state: s,
          setOptions: function (n) {
            var o = 'function' == typeof n ? n(s.options) : n
            f(),
              (s.options = Object.assign({}, i, s.options, o)),
              (s.scrollParents = {
                reference: ot(e) ? It(e) : e.contextElement ? It(e.contextElement) : [],
                popper: It(t),
              })
            var a = (function (e) {
              var t = Qt(e)
              return tt.reduce(function (e, n) {
                return e.concat(
                  t.filter(function (e) {
                    return e.phase === n
                  }),
                )
              }, [])
            })(
              (function (e) {
                var t = e.reduce(function (e, t) {
                  var n = e[t.name]
                  return (
                    (e[t.name] = n
                      ? Object.assign({}, n, t, {
                          options: Object.assign({}, n.options, t.options),
                          data: Object.assign({}, n.data, t.data),
                        })
                      : t),
                    e
                  )
                }, {})
                return Object.keys(t).map(function (e) {
                  return t[e]
                })
              })([].concat(r, s.options.modifiers)),
            )
            ;((s.orderedModifiers = a.filter(function (e) {
              return e.enabled
            })),
            (function (e) {
              e.forEach(function (t) {
                ;[]
                  .concat(Object.keys(t), tn)
                  .filter(function (e, t, n) {
                    return n.indexOf(e) === t
                  })
                  .forEach(function (n) {
                    switch (n) {
                      case 'name':
                        'string' != typeof t.name &&
                          console.error(Zt(en, String(t.name), '"name"', '"string"', '"' + String(t.name) + '"'))
                        break
                      case 'enabled':
                        'boolean' != typeof t.enabled &&
                          console.error(Zt(en, t.name, '"enabled"', '"boolean"', '"' + String(t.enabled) + '"'))
                        break
                      case 'phase':
                        tt.indexOf(t.phase) < 0 &&
                          console.error(
                            Zt(en, t.name, '"phase"', 'either ' + tt.join(', '), '"' + String(t.phase) + '"'),
                          )
                        break
                      case 'fn':
                        'function' != typeof t.fn &&
                          console.error(Zt(en, t.name, '"fn"', '"function"', '"' + String(t.fn) + '"'))
                        break
                      case 'effect':
                        null != t.effect &&
                          'function' != typeof t.effect &&
                          console.error(Zt(en, t.name, '"effect"', '"function"', '"' + String(t.fn) + '"'))
                        break
                      case 'requires':
                        null == t.requires ||
                          Array.isArray(t.requires) ||
                          console.error(Zt(en, t.name, '"requires"', '"array"', '"' + String(t.requires) + '"'))
                        break
                      case 'requiresIfExists':
                        Array.isArray(t.requiresIfExists) ||
                          console.error(
                            Zt(en, t.name, '"requiresIfExists"', '"array"', '"' + String(t.requiresIfExists) + '"'),
                          )
                        break
                      case 'options':
                      case 'data':
                        break
                      default:
                        console.error(
                          'PopperJS: an invalid property has been provided to the "' +
                            t.name +
                            '" modifier, valid properties are ' +
                            tn
                              .map(function (e) {
                                return '"' + e + '"'
                              })
                              .join(', ') +
                            '; but "' +
                            n +
                            '" was provided.',
                        )
                    }
                    t.requires &&
                      t.requires.forEach(function (n) {
                        null ==
                          e.find(function (e) {
                            return e.name === n
                          }) &&
                          console.error(
                            Zt(
                              'Popper: modifier "%s" requires "%s", but "%s" modifier is not available',
                              String(t.name),
                              n,
                              n,
                            ),
                          )
                      })
                  })
              })
            })(
              ((c = [].concat(a, s.options.modifiers)),
              (d = function (e) {
                return e.name
              }),
              (p = new Set()),
              c.filter(function (e) {
                var t = d(e)
                if (!p.has(t)) return p.add(t), !0
              })),
            ),
            ut(s.options.placement) === Je) &&
              (s.orderedModifiers.find(function (e) {
                return 'flip' === e.name
              }) ||
                console.error(
                  ['Popper: "auto" placements require the "flip" modifier be', 'present and enabled to work.'].join(
                    ' ',
                  ),
                ))
            var c,
              d,
              p,
              h = pt(t)
            return (
              [h.marginTop, h.marginRight, h.marginBottom, h.marginLeft].some(function (e) {
                return parseFloat(e)
              }) &&
                console.warn(
                  [
                    'Popper: CSS "margin" styles cannot be used to apply padding',
                    'between the popper and its reference element or boundary.',
                    'To replicate margin, use the `offset` modifier, as well as',
                    'the `padding` option in the `preventOverflow` and `flip`',
                    'modifiers.',
                  ].join(' '),
                ),
              s.orderedModifiers.forEach(function (e) {
                var t = e.name,
                  n = e.options,
                  r = void 0 === n ? {} : n,
                  o = e.effect
                if ('function' == typeof o) {
                  var i = o({ state: s, name: t, instance: l, options: r }),
                    a = function () {}
                  u.push(i || a)
                }
              }),
              l.update()
            )
          },
          forceUpdate: function () {
            if (!c) {
              var e = s.elements,
                t = e.reference,
                n = e.popper
              if (on(t, n)) {
                ;(s.rects = { reference: Kt(t, yt(n), 'fixed' === s.options.strategy), popper: ft(n) }),
                  (s.reset = !1),
                  (s.placement = s.options.placement),
                  s.orderedModifiers.forEach(function (e) {
                    return (s.modifiersData[e.name] = Object.assign({}, e.data))
                  })
                for (var r = 0, o = 0; o < s.orderedModifiers.length; o++) {
                  if ((r += 1) > 100) {
                    console.error(
                      'Popper: An infinite loop in the modifiers cycle has been detected! The cycle has been interrupted to prevent a browser crash.',
                    )
                    break
                  }
                  if (!0 !== s.reset) {
                    var i = s.orderedModifiers[o],
                      a = i.fn,
                      u = i.options,
                      f = void 0 === u ? {} : u,
                      d = i.name
                    'function' == typeof a && (s = a({ state: s, options: f, name: d, instance: l }) || s)
                  } else (s.reset = !1), (o = -1)
                }
              } else console.error(nn)
            }
          },
          update:
            ((o = function () {
              return new Promise(function (e) {
                l.forceUpdate(), e(s)
              })
            }),
            function () {
              return (
                a ||
                  (a = new Promise(function (e) {
                    Promise.resolve().then(function () {
                      ;(a = void 0), e(o())
                    })
                  })),
                a
              )
            }),
          destroy: function () {
            f(), (c = !0)
          },
        }
      if (!on(e, t)) return console.error(nn), l
      function f() {
        u.forEach(function (e) {
          return e()
        }),
          (u = [])
      }
      return (
        l.setOptions(n).then(function (e) {
          !c && n.onFirstUpdate && n.onFirstUpdate(e)
        }),
        l
      )
    }
  }
  var sn = an({
    defaultModifiers: [
      {
        name: 'eventListeners',
        enabled: !0,
        phase: 'write',
        fn: function () {},
        effect: function (e) {
          var t = e.state,
            n = e.instance,
            r = e.options,
            o = r.scroll,
            i = void 0 === o || o,
            a = r.resize,
            s = void 0 === a || a,
            u = rt(t.elements.popper),
            c = [].concat(t.scrollParents.reference, t.scrollParents.popper)
          return (
            i &&
              c.forEach(function (e) {
                e.addEventListener('scroll', n.update, jt)
              }),
            s && u.addEventListener('resize', n.update, jt),
            function () {
              i &&
                c.forEach(function (e) {
                  e.removeEventListener('scroll', n.update, jt)
                }),
                s && u.removeEventListener('resize', n.update, jt)
            }
          )
        },
        data: {},
      },
      {
        name: 'popperOffsets',
        enabled: !0,
        phase: 'read',
        fn: function (e) {
          var t = e.state,
            n = e.name
          t.modifiersData[n] = Wt({
            reference: t.rects.reference,
            element: t.rects.popper,
            strategy: 'absolute',
            placement: t.placement,
          })
        },
        data: {},
      },
      {
        name: 'computeStyles',
        enabled: !0,
        phase: 'beforeWrite',
        fn: function (e) {
          var t = e.state,
            n = e.options,
            r = n.gpuAcceleration,
            o = void 0 === r || r,
            i = n.adaptive,
            a = void 0 === i || i,
            s = n.roundOffsets,
            u = void 0 === s || s,
            c = pt(t.elements.popper).transitionProperty || ''
          a &&
            ['transform', 'top', 'right', 'bottom', 'left'].some(function (e) {
              return c.indexOf(e) >= 0
            }) &&
            console.warn(
              [
                'Popper: Detected CSS transitions on at least one of the following',
                'CSS properties: "transform", "top", "right", "bottom", "left".',
                '\n\n',
                'Disable the "computeStyles" modifier\'s `adaptive` option to allow',
                'for smooth transitions, or remove these properties from the CSS',
                'transition declaration on the popper element if only transitioning',
                'opacity or background-color for example.',
                '\n\n',
                'We recommend using the popper element as a wrapper around an inner',
                'element that can have any CSS property transitioned for animations.',
              ].join(' '),
            )
          var l = {
            placement: ut(t.placement),
            variation: Dt(t.placement),
            popper: t.elements.popper,
            popperRect: t.rects.popper,
            gpuAcceleration: o,
          }
          null != t.modifiersData.popperOffsets &&
            (t.styles.popper = Object.assign(
              {},
              t.styles.popper,
              kt(
                Object.assign({}, l, {
                  offsets: t.modifiersData.popperOffsets,
                  position: t.options.strategy,
                  adaptive: a,
                  roundOffsets: u,
                }),
              ),
            )),
            null != t.modifiersData.arrow &&
              (t.styles.arrow = Object.assign(
                {},
                t.styles.arrow,
                kt(
                  Object.assign({}, l, {
                    offsets: t.modifiersData.arrow,
                    position: 'absolute',
                    adaptive: !1,
                    roundOffsets: u,
                  }),
                ),
              )),
            (t.attributes.popper = Object.assign({}, t.attributes.popper, { 'data-popper-placement': t.placement }))
        },
        data: {},
      },
      st,
      Yt,
      Gt,
      $t,
      St,
      {
        name: 'hide',
        enabled: !0,
        phase: 'main',
        requiresIfExists: ['preventOverflow'],
        fn: function (e) {
          var t = e.state,
            n = e.name,
            r = t.rects.reference,
            o = t.rects.popper,
            i = t.modifiersData.preventOverflow,
            a = zt(t, { elementContext: 'reference' }),
            s = zt(t, { altBoundary: !0 }),
            u = Jt(a, r),
            c = Jt(s, o, i),
            l = Xt(u),
            f = Xt(c)
          ;(t.modifiersData[n] = {
            referenceClippingOffsets: u,
            popperEscapeOffsets: c,
            isReferenceHidden: l,
            hasPopperEscaped: f,
          }),
            (t.attributes.popper = Object.assign({}, t.attributes.popper, {
              'data-popper-reference-hidden': l,
              'data-popper-escaped': f,
            }))
        },
      },
    ],
  })
  He.defaults.withCredentials = !0
  var un = 'not-redraw',
    cn = function () {
      return fn(ln())
    },
    ln = function (e) {
      void 0 === e && (e = window.location)
      var t = e.search,
        n = e.hash,
        r = n.indexOf('?')
      return !t.length && r > 0 ? n.slice(r) : t
    },
    fn = function (e) {
      var t = {}
      return (
        'string' != typeof e ||
          0 === e.length ||
          (0 === e.indexOf('?') && (e = e.slice(1)),
          e.split('&').forEach(function (e) {
            var n,
              r = e.split('=')
            r.length > 0 && r[0].length > 0 && (t[r[0]] = null !== (n = r[1]) && void 0 !== n ? n : '')
          })),
        t
      )
    },
    dn = function (e) {
      if (!e) return null
      try {
        return window.document.evaluate(e, window.document, null, XPathResult.FIRST_ORDERED_NODE_TYPE, null)
          .singleNodeValue
      } catch (e) {
        return null
      }
    },
    pn = function (e, t, n, r) {
      var o = [],
        i = n()
      if (
        (t.url &&
          i &&
          (o.push(
            i[0](function () {
              ;-1 !== ln().indexOf('slardar_heatmap_draw')
                ? e()
                : o.forEach(function (e) {
                    return null == e ? void 0 : e()
                  })
            }),
          ),
          o.push(function () {
            null == r ||
              r.forEach(function (e) {
                var t,
                  n,
                  r = document.querySelector(e)
                null ===
                  (n = null === (t = null == r ? void 0 : r.parentElement) || void 0 === t ? void 0 : t.removeChild) ||
                  void 0 === n ||
                  n.call(t, r)
              })
          })),
        t.domMutation)
      ) {
        var a = new MutationObserver(function (t) {
          ;(null == t
            ? void 0
            : t.some(function (e) {
                return '' !== e.target.getAttribute(un)
              })) && e()
        })
        a.observe(window.document.body, { subtree: !0, childList: !0, attributes: !0 }),
          o.push(function () {
            var e
            return null === (e = a.disconnect) || void 0 === e ? void 0 : e.call(a)
          })
      }
      t.scroll &&
        (window.addEventListener('scroll', e, !0),
        o.push(function () {
          window.removeEventListener('scroll', e, !0)
        })),
        t.resize &&
          (window.addEventListener('resize', e, !0),
          o.push(function () {
            window.removeEventListener('resize', e, !0)
          }))
    },
    hn = (function () {
      function e(e) {
        var t = this.insertPopperDOM(e),
          n = t.domCache,
          r = t.triggerDOM,
          o = t.panelDOM
        ;(this.domCache = n), (this.triggerDOM = r), (this.panelDOM = o), (this.ins = this.initPopperJs())
      }
      return (
        (e.prototype.setText = function (e) {
          var t = this
          e.forEach(function (e) {
            var n = e.key,
              r = e.value,
              o = t.domCache[n],
              i = o.node,
              a = o.label
            i && (i.innerText = a + '' + r)
          })
        }),
        (e.prototype.insertPopperDOM = function (e) {
          var t = e.panelChildren,
            n = e.options,
            r = null != n ? n : {},
            o = r.triggerDOMClass,
            i = r.panelDOMClass,
            a = r.parentElement,
            s = document.createElement('div')
          s.setAttribute('role', 'button'),
            s.setAttribute('aria-describedby', 'tooltip'),
            s.setAttribute(un, ''),
            o && s.setAttribute('class', o)
          var u = document.createElement('div')
          u.setAttribute('role', 'tooltip'), u.setAttribute(un, ''), i && u.setAttribute('class', i)
          var c = {}
          return (
            t.forEach(function (e) {
              var t = e.key,
                n = e.label,
                r = e.className,
                o = document.createElement('p')
              o.setAttribute(un, ''), r && o.setAttribute('class', r), u.appendChild(o), (c[t] = { node: o, label: n })
            }),
            a
              ? (a.appendChild(s), a.appendChild(u))
              : (window.document.body.appendChild(s), window.document.body.appendChild(u)),
            { domCache: c, triggerDOM: s, panelDOM: u }
          )
        }),
        (e.prototype.initPopperJs = function () {
          return sn(this.triggerDOM, this.panelDOM, {
            modifiers: [Gt, $t, { name: 'preventOverflow', options: { boundary: window.document.documentElement } }],
          })
        }),
        e
      )
    })(),
    mn = { mouse: 'mouse_data_list', scroll: 'scroll_data_list' },
    vn = function () {
      return cn().slardar__heatmap__domain
    },
    gn = 'slardar__heatmap__type',
    yn = 'slardar__heatmap__id',
    bn = function (t) {
      var n = t.queryParams,
        r = t.matchListId,
        o = t.type,
        i = (function (t) {
          var n,
            r,
            o = t.queryParams,
            i = t.matchListId,
            a =
              (((n = { slardar__heatmap__bid: 'bid' })[gn] = 'heat_map_type'),
              (n.slardar__heatmap__device_type = 'device_type'),
              (n.slardar__heatmap__filter_conditions = 'filter_conditions'),
              n),
            s = fn(o),
            u = {}
          r = s[yn] ? s[yn] : Math.max.apply(Math, i)
          var c = Number(r),
            l = Number(s.slardar__heatmap__start_time),
            f = Number(s.slardar__heatmap__end_time)
          return (
            Object.keys(s).forEach(function (e) {
              if (Object.keys(a).includes(e))
                if ('slardar__heatmap__filter_conditions' === e) {
                  var t = void 0
                  try {
                    t = JSON.parse(decodeURIComponent(s[e]))
                  } catch (e) {}
                  u[a[e]] = t
                } else u[a[e]] = s[e]
            }),
            e(e({}, u), { heat_map_id: c, start_time: l, end_time: f })
          )
        })({ queryParams: n, matchListId: r })
      return He.post(
        (function (e) {
          return 'https://' + vn() + '/api_web/heat_map/' + mn[e]
        })(o),
        JSON.stringify(i),
      ).then(function (e) {
        return e.data
      })
    },
    wn = function (e) {
      var t = cn()
      return (null == t ? void 0 : t.slardar__heatmap__id)
        ? [+t.slardar__heatmap__id]
        : e
            .filter(function (e) {
              return (
                (t = e.monitor_url),
                (r = new RegExp(t, 'i')),
                !!(null === (n = r.test) || void 0 === n ? void 0 : n.call(r, location.href))
              )
              var t, n, r
            })
            .map(function (e) {
              return e.heat_map_id
            })
    },
    xn = function (e) {
      var t = document.createElement('iframe')
      ;(t.src = 'https://' + vn() + '/'),
        (t.style.width = '0px'),
        (t.style.height = '0px'),
        (t.style.borderWidth = '0px'),
        document.body.appendChild(t),
        (t.onload = function () {
          e(), document.body.removeChild(t)
        })
    },
    _n = {
      click: { pv: '', uv: '', rate: '(/)' },
      mousemove: { pv: '', uv: '', rate: '(/)' },
      mouseover: { pv: '', uv: '', rate: '(/)' },
    },
    On = {
      click: { pv: 'click pv', uv: 'click pu', rate: 'click rate(click pv/total)' },
      mousemove: { pv: 'mousemove pv', uv: 'mousemove uv', rate: 'mousemove rate(mousemove pv/total)' },
      mouseover: { pv: 'mouseover pv', uv: 'mouseover uv', rate: 'mouseover rate(mouseover pv/total)' },
    },
    En = { reach_count: '', reach_rate: '' },
    Cn = { reach_count: 'reach count', reach_rate: 'reach rate' },
    Sn = function (e) {
      return -1 !== ln().indexOf('lang=en') ? ('mouse' === e ? On : Cn) : 'mouse' === e ? _n : En
    },
    Dn = function (e) {
      var t = document.head || document.getElementsByTagName('head')[0],
        n = document.createElement('style')
      t.appendChild(n), (n.type = 'text/css'), n.appendChild(document.createTextNode(e))
    },
    An = new Map(),
    kn = { xField: 'x', yField: 'y', valueField: 'Weight' },
    jn = e(e({}, kn), { maxOpacity: 0.5, radius: 15, blur: 0.75 }),
    Tn = { click: jn, mousemove: e(e({}, kn), { radius: 5 }), mouseover: jn },
    Mn = Sn('mouse'),
    Pn = 'pv',
    Rn = 'uv',
    Ln = 'rate',
    Nn = function (e) {
      var t = e
      if (t.id) return '//*[@id="' + t.id + '"]'
      for (var n = []; t && t.nodeType === Node.ELEMENT_NODE; ) {
        for (var r = 0, o = !1, i = t.previousSibling; i; )
          i.nodeType !== Node.DOCUMENT_TYPE_NODE && i.nodeName === t.nodeName && r++, (i = i.previousSibling)
        for (i = t.nextSibling; i; ) {
          if (i.nodeName === t.nodeName) {
            o = !0
            break
          }
          i = i.nextSibling
        }
        var a = t.prefix ? t.prefix + ':' : '',
          s = r || o ? '[' + (r + 1) + ']' : ''
        n.unshift(a + t.localName + s), (t = t.parentNode)
      }
      return n.length ? '/' + n.join('/') : ''
    },
    Bn = function (e) {
      return e.replace(/\[1\]/g, '')
    },
    qn = function (e, t, n) {
      var r = [],
        o = e.length - 1,
        i = window.document.body.getBoundingClientRect(),
        a = i.top,
        s = i.left,
        u = new Map(),
        c = new MessageChannel(),
        l = c.port1,
        f = c.port2
      ;(l.onmessage = function (i) {
        var c = i.data
        if (!t() && c) {
          var l = c.index
          if (!(l > o)) {
            var d = l === o,
              p = e[l],
              h = p.xpath_name,
              m = p.points,
              v = dn(h)
            v &&
              (null == m ||
                m.some(function (e) {
                  var n = e.OffsetX,
                    o = e.OffsetY,
                    i = e.Weight,
                    c = (function (e, t, n, r, o) {
                      var i = e.getBoundingClientRect(),
                        a = i.left,
                        s = i.top,
                        u = a + window.document.body.scrollLeft - r,
                        c = s + window.document.body.scrollTop - o
                      return { x: parseInt((u + t).toFixed(), 10), y: parseInt((c + n).toFixed(), 10) }
                    })(v, n, o, s, a),
                    l = c.x,
                    f = c.y,
                    d = window.document.body,
                    p = d.scrollHeight,
                    h = d.scrollWidth
                  if (l < 0 || f < 0 || l > h || f > p) return t()
                  var m = l + '-' + f,
                    g = u.get(m)
                  if (g || 0 === g) r[g].Weight += i
                  else {
                    var y = { x: l, y: f, Weight: i }
                    r.push(y), u.set(m, r.length - 1)
                  }
                  return t()
                })),
              d ? !t() && n(r) : f.postMessage({ index: l + 1 })
          }
        }
      }),
        f.postMessage({ index: 0 })
    },
    In = function (t) {
      return t
        .map(function (t) {
          var n = t.points,
            r = t.xpath_name,
            o = (function (e, t) {
              var n = {}
              for (var r in e) Object.prototype.hasOwnProperty.call(e, r) && t.indexOf(r) < 0 && (n[r] = e[r])
              if (null != e && 'function' == typeof Object.getOwnPropertySymbols) {
                var o = 0
                for (r = Object.getOwnPropertySymbols(e); o < r.length; o++)
                  t.indexOf(r[o]) < 0 && Object.prototype.propertyIsEnumerable.call(e, r[o]) && (n[r[o]] = e[r[o]])
              }
              return n
            })(t, ['points', 'xpath_name']),
            i = dn(r)
          return e(
            {
              points: (n || [])
                .filter(function (e) {
                  return null !== e
                })
                .filter(function (e) {
                  var t = e.OffsetY,
                    n = e.OffsetX
                  if (
                    i &&
                    'function' == typeof i.getBoundingClientRect &&
                    'function' == typeof document.elementFromPoint
                  ) {
                    var r = i.getBoundingClientRect(),
                      o = r.top,
                      a = r.left,
                      s = document.elementFromPoint(a + n, o + t)
                    return s && (i === s || s.contains(i) || i.contains(s))
                  }
                }),
              xpath_name: r,
            },
            o,
          )
        })
        .filter(function (e) {
          return 0 !== e.points.length
        })
    },
    Fn = function (r, o) {
      var i = r.open_list,
        a = [],
        s = void 0,
        u = !1,
        l = null,
        d = null,
        p = void 0,
        h = !0,
        m = 200,
        v = [],
        g = !1,
        y = wn(i || []),
        b = function () {
          var e = function (e) {
            f(p) || window.clearTimeout(p),
              (p = window.setTimeout(function () {
                ;(function (e) {
                  return t(void 0, void 0, void 0, function () {
                    var t, r, o, i, a, s, u, c, l, f, p, h, m, v, g, y
                    return n(this, function (n) {
                      switch (n.label) {
                        case 0:
                          try {
                            t = e.target ? Nn(e.target) : Nn(e)
                          } catch (e) {
                            return [2]
                          }
                          return (
                            (r = e.target),
                            (o = An.get(Bn(t))),
                            d
                              ? ((i = d.triggerDOM),
                                (a = d.panelDOM),
                                o
                                  ? ((s = o.pv),
                                    (u = o.uv),
                                    (c = o.click_rate),
                                    null == i || i.setAttribute('data-show', ''),
                                    (l = r.getBoundingClientRect()),
                                    (f = l.top),
                                    (p = l.left),
                                    (h = l.bottom),
                                    (m = l.right),
                                    (v = window.document.body.getBoundingClientRect()),
                                    (g = v.top),
                                    (y = v.left),
                                    null == i ||
                                      i.setAttribute(
                                        'style',
                                        'top: ' +
                                          (f - g) +
                                          'px;left:' +
                                          (p - y) +
                                          'px;width: ' +
                                          (m - p) +
                                          'px;height: ' +
                                          (h - f) +
                                          'px',
                                      ),
                                    null == a || a.setAttribute('data-show', ''),
                                    d ? [4, d.ins.update()] : [2])
                                  : (null == i || i.removeAttribute('data-show'),
                                    null == a || a.removeAttribute('data-show'),
                                    [2]))
                              : [2]
                          )
                        case 1:
                          return (
                            n.sent(),
                            d &&
                              d.setText([
                                { key: Pn, value: s },
                                { key: Rn, value: u },
                                { key: Ln, value: (100 * c).toFixed(2) + '%' },
                              ]),
                            [2]
                          )
                      }
                    })
                  })
                })(e)
                  .then(function () {
                    p = void 0
                  })
                  .catch(function () {})
              }, 200))
          }
          v.push(function () {
            document.body.removeEventListener('mouseover', e, !0)
          }),
            document.body.addEventListener('mouseover', e, !0)
        },
        w = function () {
          bn({ queryParams: ln(), matchListId: y, type: 'mouse' })
            .then(function (t) {
              var n,
                r = t.data
              ;(n = {}),
                (a = (r || [] || []).reduce(function (t, r) {
                  var o,
                    i,
                    a,
                    s,
                    u = r.xpath_name,
                    c = r.points,
                    l = r.pv,
                    f = r.uv,
                    d = r.click_rate,
                    p = Bn(u),
                    h = n[p]
                  return (
                    h || 0 === h
                      ? (t[h] = {
                          xpath_name: p,
                          points: ((null === (o = t[h]) || void 0 === o ? void 0 : o.points) || []).concat(c),
                          pv: ((null === (i = t[h]) || void 0 === i ? void 0 : i.pv) || 0) + l,
                          uv: ((null === (a = t[h]) || void 0 === a ? void 0 : a.uv) || 0) + f,
                          click_rate: ((null === (s = t[h]) || void 0 === s ? void 0 : s.click_rate) || 0) + d,
                        })
                      : (t.push(e(e({}, r), { xpath_name: p })), (n[p] = t.length - 1)),
                    t
                  )
                }, [])),
                h
                  ? x(a)
                  : setTimeout(function () {
                      x(a)
                    }, m),
                a.forEach(function (e) {
                  An.set(e.xpath_name, e)
                })
            })
            .catch(function (e) {
              g ||
                void 0 !== e.response ||
                'Network Error' !== e.message ||
                ((g = !0),
                xn(function () {
                  w()
                }))
            })
        },
        x = function (e) {
          qn(
            In(e),
            function () {
              return u
            },
            function (e) {
              if (l) {
                l._renderer.setDimensions(0, 0)
                var t = window.document.body,
                  n = t.offsetWidth,
                  r = t.scrollWidth,
                  o = t.offsetHeight,
                  i = t.scrollHeight
                l._renderer._clear()
                var a = Math.max(n, r),
                  u = Math.max(o, i),
                  c = document.querySelector('.heatmap-draw')
                c && c.setAttribute('style', 'width: ' + a + 'px; height: ' + u + 'px'),
                  l._renderer.setDimensions(a, u),
                  l.setData({ min: 0, max: 20, data: e }),
                  (s = void 0)
              }
            },
          )
        },
        _ = function () {
          ;(u = !0),
            f(s) || window.clearTimeout(s),
            (s = window.setTimeout(function () {
              window.requestAnimationFrame(function () {
                ;(u = !1), x(a)
              })
            }, 200))
        }
      return [
        function () {
          var t,
            n = cn(),
            r = n[gn]
          Dn(
            '\nbody {\n  position: relative;\n  margin: 0;\n}\n.heatmap-draw {\n  position: absolute;\n  top: 0;\n  left: 0;\n  bottom: 0;\n  right: 0;\n  pointer-events: none;\n}\n\n.heatmap-draw__mask {\n  position: absolute;\n  top: 0;\n  left: 0;\n  width: 100%;\n  height: 100%;\n  background-color: #000;\n  opacity: 0.2;\n  z-index: 9991;\n}\n\n.heatmap-draw-canvas-container {\n  position: absolute;\n  top: 0;\n  width: 100%;\n  height: 100%;\n  z-index: 9994;\n}\n\n#drawContainer {\n  width: 100%;\n  height: 100%;\n}\n\n.heatmap-selected-dom {\n  border: 1px solid #fff;\n  position: absolute;\n  z-index: 9992;\n  opacity: 0;\n}\n\n.heatmap-selected-dom[data-show] {\n  opacity: 1;\n}\n\n.heatmap-selected-data-panel {\n  background: #475fd5;\n  color: #fff;\n  padding: 12px;\n  max-width: 300px;\n  min-width: 150px;\n  z-index: 9998;\n  opacity: 0;\n}\n\n.heatmap-selected-data-panel[data-show] {\n  opacity: 1;\n}\n',
          ),
            (t = document.createElement('div')).setAttribute('class', 'heatmap-draw'),
            t.setAttribute(un, ''),
            (t.innerHTML =
              '\n  <div class="heatmap-draw__mask" not-redraw></div>\n  <div class="heatmap-draw-canvas-container" not-redraw>\n    <div id="drawContainer" not-redraw></div>\n  </div>\n  '),
            (document.body || document.getElementsByTagName('body')[0]).appendChild(t),
            (function (t) {
              var n,
                r = document.getElementById('drawContainer')
              if (r) {
                var o
                try {
                  o = JSON.parse(localStorage.getItem('heat_map_config')) || {}
                } catch (e) {
                  o = {}
                }
                ;(l = c.create(e(e(e({}, Tn[t]), { container: r }), o))),
                  null === (n = window.document.querySelector('.heatmap-canvas')) ||
                    void 0 === n ||
                    n.setAttribute(un, '')
              }
            })(r),
            (function (e) {
              d = new hn({
                panelChildren: [
                  { key: Pn, label: Mn[e].pv },
                  { key: Rn, label: Mn[e].uv },
                  { key: Ln, label: Mn[e].rate },
                ],
                options: {
                  triggerDOMClass: 'heatmap-selected-dom',
                  panelDOMClass: 'heatmap-selected-data-panel',
                  parentElement: window.document.querySelector('.heatmap-draw'),
                },
              })
            })(r),
            b(),
            (h = 'false' !== n.auto_refresh)
          var i = Number(n.refresh_timeout)
          ;(m = Number.isNaN(i) ? 200 : i),
            w(),
            h && pn(_, { scroll: !0, domMutation: !0, url: !0, resize: !0 }, o, ['.heatmap-draw'])
        },
        function () {
          v.forEach(function (e) {
            return e()
          }),
            (v = [])
        },
      ]
    }
  function Un() {
    if ('object' == typeof window && l(window)) return window
  }
  function Hn() {
    if ('object' == typeof document && l(document)) return document
  }
  var Wn = 'reach_count',
    zn = 'reach_rate',
    Vn = Sn('scroll'),
    Gn = {
      getGradientColors: function (e, t) {
        for (var n = [], r = 0, o = e; r < o.length; r++) {
          var i = o[r]
          i.indexOf('rgb') >= 0 ? n.push(this.rgbTo16(i)) : n.push(i)
        }
        return this.gradientColor(n, t)
      },
      rgbTo16: function (e) {
        var t = e
            .replace(/rgba?\(/, '')
            .replace(/\)/, '')
            .replace(/[\s+]/g, '')
            .split(','),
          n = parseFloat(t[3] || '1'),
          r = Math.floor(n * parseInt(t[0]) + 255 * (1 - n)),
          o = Math.floor(n * parseInt(t[1]) + 255 * (1 - n)),
          i = Math.floor(n * parseInt(t[2]) + 255 * (1 - n))
        return (
          '#' + ('0' + r.toString(16)).slice(-2) + ('0' + o.toString(16)).slice(-2) + ('0' + i.toString(16)).slice(-2)
        )
      },
      gradientColor: function (e, t) {
        var n = this.toRgb(e)
        t -= 1
        for (
          var r = Math.floor(t / (n.length - 1)), o = t - r * (n.length - 1), i = 0, a = [], s = 0;
          s < n.length - 1;
          s++
        ) {
          var u = n[s].r,
            c = n[s].g,
            l = n[s].b,
            f = n[s + 1].r,
            d = n[s + 1].g,
            p = n[s + 1].b
          ;(i = r), o > 0 && ((i = r + 1), o--)
          for (var h = (f - u) / i, m = (d - c) / i, v = (p - l) / i, g = 0; g < i; g++) {
            var y = this.colorHex(
              'rgb(' + parseInt(h * g + u + '') + ',' + parseInt(m * g + c + '') + ',' + parseInt(v * g + l + '') + ')',
            )
            a.push(y)
          }
          if (s === n.length - 2) {
            y = 'rgb(' + f + ',' + d + ',' + p + ')'
            a.push(y)
          }
        }
        return a
      },
      toRgb: function (e) {
        var t = this
        return e.map(function (e) {
          var n = t.colorRgb(e)
          return { r: n[0], g: n[1], b: n[2] }
        })
      },
      colorRgb: function (e) {
        var t = e.toLowerCase()
        if (t && /^#([0-9a-fA-f]{3}|[0-9a-fA-f]{6})$/.test(t)) {
          if (4 === t.length) {
            for (var n = '#', r = 1; r < 4; r += 1) n += t.slice(r, r + 1).concat(t.slice(r, r + 1))
            t = n
          }
          var o = []
          for (r = 1; r < 7; r += 2) o.push(parseInt('0x' + t.slice(r, r + 2)))
          return o
        }
        return []
      },
      colorHex: function (e) {
        var t = e
        if (/^(rgb|RGB)/.test(t)) {
          for (var n = '#', r = 0, o = t.replace(/(?:(|)|rgb|RGB)*/g, '').split(','); r < o.length; r++) {
            var i = o[r],
              a = Number(i).toString(16)
            '0' === (a = parseInt(a) < 10 ? '0' + a : a) && (a += a), (n += a)
          }
          return 7 !== n.length && (n = t), n
        }
        if (/^#([0-9a-fA-f]{3}|[0-9a-fA-f]{6})$/.test(t)) {
          var s = t.replace(/#/, '').split('')
          if (6 === s.length) return t
          if (3 === s.length)
            return s.reduce(function (e, t) {
              return '' + e + t + t
            }, '#')
        }
        return t
      },
    },
    Jn = function (e, t) {
      var n,
        r,
        o,
        i,
        a,
        s = e.getBoundingClientRect().width,
        u = e.scrollHeight,
        c = e.scrollWidth,
        l = document.createElement('div'),
        f = document.createElement('canvas'),
        d = ((i = [s, c]), (a = 30), Math.abs(i[1] - i[0]) > a ? i[1] : i[0]),
        p = u
      return (
        e &&
          ['', 'static'].indexOf(
            null !== (r = null === (n = null == e ? void 0 : e.style) || void 0 === n ? void 0 : n.position) &&
              void 0 !== r
              ? r
              : '',
          ) > -1 &&
          (e.style.position = 'relative'),
        (l.id = 'heatmap_cover'),
        (l.style.width = d + 'px'),
        (l.style.height = p + 'px'),
        (l.style.zIndex = '9999'),
        (l.style.position = 'absolute'),
        (l.style.top = '0'),
        (l.style.border = '1px solid'),
        (f.width = d),
        (f.height = p),
        e.appendChild(l),
        l.appendChild(f),
        t && (null === (o = e.removeChild) || void 0 === o || o.call(e, t)),
        e.setAttribute(un, ''),
        l.setAttribute(un, ''),
        [f, l]
      )
    },
    Xn = function (e, o) {
      var i = e.open_list,
        a = null,
        s = void 0,
        u = void 0,
        c = void 0,
        l = wn(i || []),
        d = null,
        p = null,
        h = [],
        m = !1,
        v = function (e) {
          f(s) || window.clearTimeout(s),
            (s = window.setTimeout(function () {
              ;(function (e) {
                return t(void 0, void 0, void 0, function () {
                  var t, r, o, i, s, u, l
                  return n(this, function (n) {
                    switch (n.label) {
                      case 0:
                        return c
                          ? ((r = (t = c).renderList),
                            (o = t.max),
                            (i = Math.floor(e.offsetY)),
                            (s = r[i]),
                            a
                              ? ((u = a.triggerDOM),
                                (l = a.panelDOM),
                                s
                                  ? ((u.style.width = '100%'),
                                    (u.style.borderTop = '1px solid #333'),
                                    (u.style.position = 'absolute'),
                                    (u.style.top = i + 'px'),
                                    null == l || l.setAttribute('data-show', ''),
                                    null == u || u.setAttribute('data-show', ''),
                                    [4, a.ins.update()])
                                  : (null == u || u.removeAttribute('data-show'),
                                    null == l || l.removeAttribute('data-show'),
                                    [2]))
                              : [2])
                          : [2]
                      case 1:
                        return (
                          n.sent(),
                          a.setText([
                            { key: Wn, value: '' + s },
                            { key: zn, value: Math.floor((100 * s) / o) + '%' },
                          ]),
                          [2]
                        )
                    }
                  })
                })
              })(e)
                .then(function () {
                  s = void 0
                })
                .catch(function () {})
            }, 500))
        },
        g = function (e) {
          !(function (e) {
            a = new hn({
              panelChildren: [
                { key: Wn, label: Vn.reach_count },
                { key: zn, label: Vn.reach_rate },
              ],
              options: {
                triggerDOMClass: 'heatmap-selected-dom',
                panelDOMClass: 'heatmap-selected-data-panel',
                parentElement: e.parentElement,
              },
            })
          })(e)
          var t = function (e) {
            v(e)
          }
          h.push(function () {
            e.removeEventListener('mousemove', t)
          }),
            e.addEventListener('mousemove', t)
        },
        y = function () {
          var e = c,
            t = e.node,
            n = e.max,
            o = e.min,
            i = e.renderList,
            a = void 0 === i ? [] : i,
            s = Jn(t, p),
            u = s[0],
            l = s[1]
          p = l
          var f = (function (e, t, n) {
            var o = t + 1,
              i = Gn.getGradientColors(['#000', '#00F', '#0FF', '#FF0', '#F00'], o)
            return e
              .map(function (e, t) {
                return { color: i[e - n], viewCount: e, y: t }
              })
              .reduce(function (t, n, o) {
                return n.viewCount === e[o - 1] ? t : r(r([], t, !0), [{ y: n.y, color: n.color }], !1)
              }, [])
          })(a, n, o)
          g(u),
            (function (e, t, n) {
              if (e.getContext) {
                var r = e.getContext('2d')
                if (r) {
                  r.globalAlpha = 0.6
                  for (var o = 0, i = 0, a = n; i < a.length; i++) {
                    var s = a[i],
                      u = s.y,
                      c = s.color
                    r.beginPath(), (r.fillStyle = c), r.fillRect(0, o, t, u - o), (o = u), r.fill(), r.closePath()
                  }
                }
              }
            })(u, u.width, f)
        },
        b = function () {
          if (d) {
            var e,
              t,
              n,
              r =
                ((e = d),
                (Array.isArray(null == e ? void 0 : e.data) ? e.data : []).reduce(function (e, t) {
                  var n, r, o, i, a, s
                  if (!dn(t.xpath_name)) return e
                  var u =
                      null !==
                        (r = null === (n = null == e ? void 0 : e.visit_count) || void 0 === n ? void 0 : n.length) &&
                      void 0 !== r
                        ? r
                        : 0,
                    c =
                      null !==
                        (i = null === (o = null == t ? void 0 : t.visit_count) || void 0 === o ? void 0 : o.length) &&
                      void 0 !== i
                        ? i
                        : 0,
                    l = null !== (a = null == e ? void 0 : e.max_count) && void 0 !== a ? a : 0,
                    f = null !== (s = null == t ? void 0 : t.max_count) && void 0 !== s ? s : 0
                  return f > l || (f === l && c > u) ? t : e
                }, {})),
              o = (function (e) {
                var t,
                  n,
                  r = e.visit_count,
                  o = e.xpath_name,
                  i = e.max_count,
                  a = e.min_count,
                  s = dn(o)
                if (s) {
                  var u = null !== (t = r.length) && void 0 !== t ? t : 0,
                    c = null !== (n = s.scrollHeight) && void 0 !== n ? n : u
                  return {
                    max: i,
                    min: a,
                    node: s,
                    renderList: c <= u ? r : r.concat(new Array(c - u).fill(0)),
                    nodeHeight: s.scrollHeight,
                  }
                }
              })(r)
            if (o) {
              var i = c
              ;(c = o),
                (t = o),
                (f((n = i)) ||
                  t.max !== n.max ||
                  t.min !== n.min ||
                  t.nodeHeight !== n.nodeHeight ||
                  t.node !== n.node ||
                  t.renderList.length !== n.renderList.length ||
                  !t.renderList.every(function (e, t) {
                    return e !== n.renderList[t]
                  })) &&
                  y()
            }
          }
        },
        w = function () {
          bn({ queryParams: ln(), matchListId: l, type: 'scroll' })
            .then(function (e) {
              ;(d = e),
                (function (e) {
                  var t,
                    n,
                    r,
                    o = Un(),
                    i = Hn()
                  o &&
                    i &&
                    ('complete' !== i.readyState
                      ? ((t = 'load'),
                        (n = function () {
                          setTimeout(function () {
                            e()
                          }, 0)
                        }),
                        void 0 === (r = !1) && (r = !1),
                        o.addEventListener(t, n, r))
                      : e())
                })(function () {
                  b()
                })
            })
            .catch(function (e) {
              m ||
                void 0 !== e.response ||
                'Network Error' !== e.message ||
                ((m = !0),
                xn(function () {
                  w()
                }))
            })
        },
        x = function () {
          f(u) || window.clearTimeout(u),
            (u = window.setTimeout(function () {
              window.requestAnimationFrame(function () {
                b(), (u = void 0)
              })
            }, 500))
        }
      return [
        function () {
          l.length &&
            (Dn(
              '\nbody {\n  position: relative;\n  margin: 0;\n}\n.heatmap-selected-dom {\n  width: 100%;\n  border-top: 1px solid rgb(51, 51, 51);\n  top: -10000px;\n  position: absolute;\n  z-index: 3;\n  opacity: 0;\n}\n\n.heatmap-selected-dom[data-show] {\n  opacity: 1;\n}\n\n.heatmap-selected-data-panel {\n  background: #475fd5;\n  color: #fff;\n  padding: 12px;\n  max-width: 300px;\n  min-width: 150px;\n  z-index: 8;\n  opacity: 0;\n}\n\n.heatmap-selected-data-panel[data-show] {\n  opacity: 1;\n}\n',
            ),
            w(),
            pn(x, { scroll: !0, domMutation: !0, url: !0, resize: !0 }, o, ['#heatmap_cover']))
        },
        function () {
          h.forEach(function (e) {
            return e()
          }),
            (h = [])
        },
      ]
    },
    Yn = ['click', 'mouseover', 'mousemove'],
    $n = ['scroll']
  window.drawHeatmap = function (e, t) {
    var n = cn()[gn]
    if (Yn.includes(n)) {
      var r = Fn(e, t),
        o = r[0],
        i = r[1]
      return o(), i
    }
    if ($n.includes(n)) {
      var a = Xn(e, t)
      ;(o = a[0]), (i = a[1])
      return o(), i
    }
  }
})()
