// @ai-local-server/common/dist/configCenter/index.js
import { __decorate } from "tslib";
import 'reflect-metadata';
import { ConfigCenterCore } from './configCenterCore';
import { SingletonInjectable } from '../common/injection';
export * from './configCenterCore';
let ConfigCenter = class ConfigCenter extends ConfigCenterCore {
    getAIIntents(defaultValue) {
        return this.get('aiIntents', defaultValue);
    }
    getAISlashCommands(defaultValue) {
        return this.get('aiSlashCommands', defaultValue);
    }
    getAISideChatConfig(defaultValue) {
        return this.get('aiSideChatConfig', defaultValue);
    }
    getAIWelcomeConfig(defaultValue) {
        return this.get('aiWelcomeConfig', defaultValue);
    }
    getAILoginConfig(defaultValue) {
        return this.get('aiLoginConfig', defaultValue);
    }
    getAIAdditionalResolvers(defaultValue) {
        return this.get('aiAdditionalResolvers', defaultValue);
    }
    getFeatureGates(defaultValue) {
        return this.get('featureGates', defaultValue);
    }
    getTeaChannel(defaultValue) {
        return this.get('teaChannel', defaultValue);
    }
    getAIHashContexts(defaultValue) {
        return this.get('aiHashContexts', defaultValue);
    }
};
ConfigCenter = __decorate([
    SingletonInjectable
], ConfigCenter);
export { ConfigCenter };
