// @slardar/sdk-web/esm/collector/dom.js
import { getDefaultDocument } from '../utils/defaults';
import { addListenerToDocument } from '../utils/misc';
export var CLICK_EVENT_NAME = 'click';
export var CLICK_SUBJECT_NAME = CLICK_EVENT_NAME + "_0";
export var DOM_EVENT_TYPE = 'dom';
export var observeClick = function (next, tearDown) {
    var document = getDefaultDocument();
    if (!document)
        return;
    var removeListener = addListenerToDocument(document, CLICK_EVENT_NAME, next, true);
    tearDown(function () {
        removeListener();
    });
};
export var clickSubject = [CLICK_SUBJECT_NAME, observeClick];
export var KEYPRESS_EVENT_NAME = 'keypress';
export var KEYPRESS_SUBJECT_NAME = KEYPRESS_EVENT_NAME + "_0";
export var observeKeypress = function (next, tearDown) {
    var document = getDefaultDocument();
    if (!document)
        return;
    var removeListener = addListenerToDocument(document, KEYPRESS_EVENT_NAME, next, true);
    tearDown(function () {
        removeListener();
    });
};
export var keypressSubject = [KEYPRESS_SUBJECT_NAME, observeKeypress];//# sourceMappingURL=https://trae.private/sourcemaps/424b4bd987c6c6a4cadf1a08da420593cc6bf75a/node_modules/@byted-icube/devtools-base/node_modules/@slardar/sdk-web/esm/collector/dom.js.map
