// @slardar/web/esm/__test__/util.js
import { createBaseConfigManager, createClient, directBuilder, id, createBatchSender, noop, } from '@slardar/sdk-template';
import { getXhrTransport, now } from '@slardar/sdk-web';
export var createTestWebClient = function () {
    return createClient({
        builder: directBuilder,
        createSender: function () {
            return createBatchSender({
                endpoint: 'endpoint',
                transport: {
                    get: noop,
                    post: noop,
                },
                wait: 1000,
            });
        },
        createDefaultConfig: function () { return ({
            bid: '',
            pid: '',
            viewId: '_' + now(),
            userId: '',
            deviceId: '',
            sessionId: '',
            domain: '',
            pluginPathPrefix: '',
            plugins: {},
            sample: {
                sample_rate: 1,
                sample_granularity: 'session',
                include_users: [],
                rules: {},
            },
            env: '',
            release: '',
            transport: getXhrTransport(),
        }); },
        userConfigNormalizer: id,
        initConfigNormalizer: id,
        createConfigManager: createBaseConfigManager,
        validateInitConfig: function () { return true; },
    });
};
export var proxyLocalStorage = function () {
    var localStorageMock = (function () {
        var store = {};
        return {
            getItem: function (key) {
                return store[key];
            },
            setItem: function (key, value) {
                store[key] = value.toString();
            },
            clear: function () {
                store = {};
            },
            removeItem: function (key) {
                delete store[key];
            },
        };
    })();
    Object.defineProperty(global, 'localStorage', { value: localStorageMock });
};//# sourceMappingURL=https://trae.private/sourcemaps/424b4bd987c6c6a4cadf1a08da420593cc6bf75a/node_modules/@slardar/integrations/node_modules/@slardar/web/esm/__test__/util.js.map
