// @slardar/web/esm/constants.euttp.js
var REPORT_DOMAIN = "mon-i18n.tiktokv.com";
var SETTINGS_DOMAIN = REPORT_DOMAIN;
var PLUGINS_LOAD_PREFIX = "https:" + "//" + ("sf16-short-va.bytedapm.com") + "/slardar/fe/sdk-web/plugins";
var SDK_VERSION = "1.12.3" ;
var SDK_NAME = 'SDK_SLARDAR_WEB';
var SETTINGS_PATH = '/monitor_web/settings/browser-settings';
var BATCH_REPORT_PATH = "/monitor_browser/collect/batch/";
var STORAGE_PREFIX = 'SLARDAR';
var USER_ID_COOKIE_NAME = 'MONITOR_WEB_ID';
var DEVICE_ID_COOKIE_NAME = 'MONITOR_DEVICE_ID';
var DEFAULT_IGNORE_PATHS = ['/log/sentry/', BATCH_REPORT_PATH, SETTINGS_PATH];
var DEFAULT_SAMPLE_GRANULARITY = 'session';
var EXTRA_INDEPENDENT_PLUGINS = ['blankScreen', 'action'];
var DEFAULT_SAMPLE_CONFIG = {
    sample_rate: 1,
    include_users: [],
    sample_granularity: DEFAULT_SAMPLE_GRANULARITY,
    rules: {},
};
var DEFAULT_SENDER_SIZE = 20;

export { BATCH_REPORT_PATH, DEFAULT_IGNORE_PATHS, DEFAULT_SAMPLE_CONFIG, DEFAULT_SAMPLE_GRANULARITY, DEFAULT_SENDER_SIZE, DEVICE_ID_COOKIE_NAME, EXTRA_INDEPENDENT_PLUGINS, PLUGINS_LOAD_PREFIX, REPORT_DOMAIN, SDK_NAME, SDK_VERSION, SETTINGS_DOMAIN, SETTINGS_PATH, STORAGE_PREFIX, USER_ID_COOKIE_NAME };
