// @slardar/sdk-web/esm/monitors/__test__/utils.js
import { HTTP_EV_TYPE, RESOURCE_ERROR_EV_TYPE, RESOURCE_EV_TYPE } from '../../constants';
export function mockHttpEvent(url) {
    var originalUrl = url || 'http://example.com/123';
    var timing = {
        name: originalUrl,
        toJSON: function () {
            return {
                name: originalUrl,
            };
        },
    };
    var httpPayload = {
        api: 'xhr',
        request: {
            url: originalUrl,
            method: 'get',
            timestamp: 1,
        },
        response: {
            status: 200,
            is_custom_error: false,
            timestamp: 2,
            timing: timing,
        },
        duration: 3,
    };
    return {
        ev_type: HTTP_EV_TYPE,
        payload: httpPayload,
    };
}
export function mockResourceEvent(url) {
    var originalUrl = url || 'http://example.com/123';
    var resourcePayload = {
        name: originalUrl,
        entryType: 'resource',
        startTime: 949.0999999046326,
        duration: 0,
        initiatorType: 'link',
        deliveryType: 'cache',
        nextHopProtocol: 'h2',
        renderBlockingStatus: 'non-blocking',
        workerStart: 0,
        redirectStart: 0,
        redirectEnd: 0,
        fetchStart: 949.0999999046326,
        domainLookupStart: 949.0999999046326,
        domainLookupEnd: 949.0999999046326,
        connectStart: 949.0999999046326,
        secureConnectionStart: 949.0999999046326,
        connectEnd: 949.0999999046326,
        requestStart: 949.0999999046326,
        responseStart: 949.0999999046326,
        firstInterimResponseStart: 0,
        responseEnd: 949.0999999046326,
        transferSize: 0,
        encodedBodySize: 324864,
        decodedBodySize: 324864,
        responseStatus: 200,
        serverTiming: [],
    };
    return {
        ev_type: RESOURCE_EV_TYPE,
        payload: resourcePayload,
    };
}
export function mockResourceErrorEvent(url) {
    var originalUrl = url || 'http://example.com/123';
    return {
        ev_type: RESOURCE_ERROR_EV_TYPE,
        payload: {
            type: 'script',
            url: originalUrl,
        },
    };
}//# sourceMappingURL=https://trae.private/sourcemaps/424b4bd987c6c6a4cadf1a08da420593cc6bf75a/node_modules/@byted-icube/devtools-base/node_modules/@slardar/sdk-web/esm/monitors/__test__/utils.js.map
