// @slardar/sdk-web/esm/collector/error.js
import { getDefaultBrowser } from '../utils/defaults';
export var ERROR_SUBJECT_NAME = 'err_0';
export var observeError = function (next, tearDown) {
    var window = getDefaultBrowser();
    if (!window)
        return;
    window.addEventListener('error', next, true);
    tearDown(function () {
        window.removeEventListener('error', next, true);
    });
};
export var errorSubject = [ERROR_SUBJECT_NAME, observeError];
export var PROMISE_REJECTION_SUBJECT_NAME = 'perr_0';
export var observePromiseRejection = function (next, tearDown) {
    var window = getDefaultBrowser();
    if (!window)
        return;
    window.addEventListener('unhandledrejection', next, true);
    tearDown(function () {
        window.removeEventListener('unhandledrejection', next, true);
    });
};
export var promiseRejectionSubject = [
    PROMISE_REJECTION_SUBJECT_NAME,
    observePromiseRejection,
];//# sourceMappingURL=https://trae.private/sourcemaps/424b4bd987c6c6a4cadf1a08da420593cc6bf75a/node_modules/@slardar/integrations/node_modules/@slardar/sdk-web/esm/collector/error.js.map
