// @byted-icube/devtools-base/dist/constants/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AUTH_BASE_URL_MAP = exports.DEFAULT_ELECTRON_OVERSEA_SLARDAR_BID = exports.DEFAULT_ELECTRON_SLARDAR_BID = exports.DEFAULT_SLARDAR_BID_MAP = exports.TEA_AB_CHANNEL_DOMAIN_MAP = exports.TEA_AB_SERVICE_ORIGIN_MAP = exports.TEA_CHANNEL_MAP = exports.TEA_OVERSEA_ELECTRON_APP_ID = exports.TEA_ELECTRON_APP_ID = exports.TEA_APP_ID_MAP = exports.DEFAULT_REGION = exports.DEFAULT_STARLING_FALLBACK_LANG = exports.DEFAULT_STARLING_LANG = exports.DEFAULT_STARLING_NAMESPACE = exports.DEFAULT_STARLING_APP_KEY = void 0;
const types_1 = require("../types");
exports.DEFAULT_STARLING_APP_KEY = '4b68ec50ae9d11ee9b92612267a029f0';
exports.DEFAULT_STARLING_NAMESPACE = 'devtools';
exports.DEFAULT_STARLING_LANG = 'en';
exports.DEFAULT_STARLING_FALLBACK_LANG = ['en', 'zh'];
exports.DEFAULT_REGION = types_1.Region.CN;
// https://bytedance.larkoffice.com/wiki/XeXKwQpnTiblcTkAL2ac3Oi7nWb
exports.TEA_APP_ID_MAP = {
    [types_1.Region.CN]: {
        inner: 379586,
        outer: 576092,
    },
    [types_1.Region.BOE]: {
        inner: 379586,
        outer: 576092,
    },
    [types_1.Region.US]: {
        inner: 379586,
        outer: 573425,
    },
    [types_1.Region.SG]: {
        inner: 379586,
        outer: 573425,
    },
    [types_1.Region.BOEI18N]: {
        inner: 379586,
        outer: 573425,
    },
};
exports.TEA_ELECTRON_APP_ID = 647416; // 目前只有cn有electron版本
exports.TEA_OVERSEA_ELECTRON_APP_ID = 677332;
exports.TEA_CHANNEL_MAP = {
    [types_1.Region.CN]: {
        inner: 'cn',
        outer: 'cn',
    },
    [types_1.Region.BOE]: {
        inner: 'cn',
        outer: 'cn',
    },
    [types_1.Region.US]: {
        inner: 'cn',
        outer: 'va',
    },
    [types_1.Region.SG]: {
        inner: 'cn',
        outer: 'sg',
    },
    [types_1.Region.BOEI18N]: {
        inner: 'cn',
        outer: 'sg',
    },
};
exports.TEA_AB_SERVICE_ORIGIN_MAP = {
    [types_1.Region.CN]: {
        inner: 'https://abtestvm.bytedance.com',
        outer: 'https://abtestvm.bytedance.com',
    },
    [types_1.Region.BOE]: {
        inner: 'https://abtestvm.bytedance.com',
        outer: 'https://abtestvm.bytedance.com',
    },
    [types_1.Region.US]: {
        inner: 'https://abtestvm.bytedance.com',
        outer: 'https://libraweb-va.tiktok.com',
    },
    [types_1.Region.SG]: {
        inner: 'https://abtestvm.bytedance.com',
        outer: 'https://libraweb-sg.tiktok.com',
    },
    [types_1.Region.BOEI18N]: {
        inner: 'https://abtestvm.bytedance.com',
        outer: 'https://libraweb-sg.tiktok.com',
    },
};
// https://bytedance.larkoffice.com/wiki/wikcnSZCeeUVs4B4CdRfa8ghoXb
// cn: "https://abtestvm.bytedance.com"
// sg: "https://libraweb-sg.tiktok.com"
// gcp: "https://libraweb-i18n.tiktok.com"
// va: "https://libraweb-va.tiktok.com"
// us-ttp & us-ttp2: "https://libraweb.tiktokw.us"
// eu-ttp: "https://libraweb.tiktokw.eu"
exports.TEA_AB_CHANNEL_DOMAIN_MAP = {
    [types_1.Region.CN]: {
        inner: 'https://abtestvm.bytedance.com',
        outer: 'https://abtestvm.bytedance.com',
    },
    [types_1.Region.BOE]: {
        inner: 'https://abtestvm.bytedance.com',
        outer: 'https://abtestvm.bytedance.com',
    },
    [types_1.Region.US]: {
        inner: 'https://abtestvm.bytedance.com',
        outer: 'https://libraweb-va.tiktok.com',
    },
    [types_1.Region.SG]: {
        inner: 'https://abtestvm.bytedance.com',
        outer: 'https://libraweb-sg.tiktok.com',
    },
    [types_1.Region.BOEI18N]: {
        inner: 'https://abtestvm.bytedance.com',
        outer: 'https://libraweb-sg.tiktok.com',
    },
};
exports.DEFAULT_SLARDAR_BID_MAP = {
    [types_1.Region.CN]: {
        inner: 'icube_web',
        outer: 'stable_ide_cn',
    },
    [types_1.Region.BOE]: {
        inner: 'icube_web',
        outer: 'stable_ide_cn',
    },
    [types_1.Region.US]: {
        inner: 'cloud_ide_us',
        outer: 'stable_ide_us',
    },
    [types_1.Region.SG]: {
        inner: 'cloud_ide_us',
        outer: 'stable_ide_us',
    },
    [types_1.Region.BOEI18N]: {
        inner: 'cloud_ide_us',
        outer: 'stable_ide_us',
    },
};
exports.DEFAULT_ELECTRON_SLARDAR_BID = 'marscode_native_ide_cn'; // TODO 目前只有 cn, 后续去掉
exports.DEFAULT_ELECTRON_OVERSEA_SLARDAR_BID = 'marscode_native_ide_us'; // TODO 目前只有 cn, 后续去掉
exports.AUTH_BASE_URL_MAP = {
    [types_1.Region.CN]: '',
    [types_1.Region.BOE]: '',
    [types_1.Region.US]: '',
    [types_1.Region.SG]: '',
    [types_1.Region.BOEI18N]: '',
};
