// @slardar/web/sg/constants.sg.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var REPORT_DOMAIN = "mon-va.byteoversea.com";
var SETTINGS_DOMAIN = REPORT_DOMAIN;
var PLUGINS_LOAD_PREFIX = "https:" + "//" + ("sf16-short-sg.bytedapm.com") + "/slardar/fe/sdk-web/plugins";
var SDK_VERSION = "1.14.1" ;
var SDK_NAME = 'SDK_SLARDAR_WEB';
var SETTINGS_PATH = '/monitor_web/settings/browser-settings';
var BATCH_REPORT_PATH = "/monitor_browser/collect/batch/";
var STORAGE_PREFIX = 'SLARDAR';
var USER_ID_COOKIE_NAME = 'MONITOR_WEB_ID';
var DEVICE_ID_COOKIE_NAME = 'MONITOR_DEVICE_ID';
var DEFAULT_IGNORE_PATHS = ['/log/sentry/', BATCH_REPORT_PATH, SETTINGS_PATH];
var DEFAULT_SAMPLE_GRANULARITY = 'session';
var EXTRA_INDEPENDENT_PLUGINS = ['blankScreen', 'action'];
var DEFAULT_SAMPLE_CONFIG = {
    sample_rate: 1,
    include_users: [],
    sample_granularity: DEFAULT_SAMPLE_GRANULARITY,
    rules: {},
};
var DEFAULT_SENDER_SIZE = 20;

exports.BATCH_REPORT_PATH = BATCH_REPORT_PATH;
exports.DEFAULT_IGNORE_PATHS = DEFAULT_IGNORE_PATHS;
exports.DEFAULT_SAMPLE_CONFIG = DEFAULT_SAMPLE_CONFIG;
exports.DEFAULT_SAMPLE_GRANULARITY = DEFAULT_SAMPLE_GRANULARITY;
exports.DEFAULT_SENDER_SIZE = DEFAULT_SENDER_SIZE;
exports.DEVICE_ID_COOKIE_NAME = DEVICE_ID_COOKIE_NAME;
exports.EXTRA_INDEPENDENT_PLUGINS = EXTRA_INDEPENDENT_PLUGINS;
exports.PLUGINS_LOAD_PREFIX = PLUGINS_LOAD_PREFIX;
exports.REPORT_DOMAIN = REPORT_DOMAIN;
exports.SDK_NAME = SDK_NAME;
exports.SDK_VERSION = SDK_VERSION;
exports.SETTINGS_DOMAIN = SETTINGS_DOMAIN;
exports.SETTINGS_PATH = SETTINGS_PATH;
exports.STORAGE_PREFIX = STORAGE_PREFIX;
exports.USER_ID_COOKIE_NAME = USER_ID_COOKIE_NAME;
