// @ies/tron-client/dist/core/logger.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const chalk_1 = __importDefault(require("chalk"));
class Logger {
    constructor() {
        this._logger = console;
    }
    /**
     * setup logger
     * @param logger custom logger such as electron log
     */
    setup(logger) {
        this._logger = logger;
    }
    /**
     * disable logger
     */
    disable() {
        // prettier-ignore
        /* eslint-disable @typescript-eslint/no-empty-function */
        this._logger = {
            info: () => { },
            error: () => { },
            warn: () => { },
            log: () => { }
        };
    }
    info(...message) {
        this._logger.info(chalk_1.default.blue('[TRON CLIENT]: INFO '), ...message);
    }
    error(...message) {
        this._logger.error(chalk_1.default.red('[TRON CLIENT]: ERROR'), ...message);
    }
    warn(...message) {
        this._logger.warn(chalk_1.default.yellow('[TRON CLIENT]: WARN '), ...message);
    }
    log(...message) {
        this._logger.log(chalk_1.default.green('[TRON CLIENT]: LOG  '), ...message);
    }
}
exports.default = new Logger();//# sourceMappingURL=https://trae.private/sourcemaps/424b4bd987c6c6a4cadf1a08da420593cc6bf75a/node_modules/@ies/tron-client/dist/core/logger.js.map
