// @ies/tron-client/dist/core/tea.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.sendEvent = void 0;
const tea_sdk_node_1 = require("@dp/tea-sdk-node");
const node_plugin_http_1 = require("@logsdk/node-plugin-http");
const teaSDK = new tea_sdk_node_1.TeaSDK({
    app_id: 4120,
    logLevel: 'error',
});
const httpPluginController = (0, node_plugin_http_1.httpPlugin)({
    channel: 'cn',
    event_verify_url: 'https://mcs.bytedance.net/v1/list_test',
    retry: 5,
});
teaSDK.use(httpPluginController);
function sendEvent(eventName, param, header = {}) {
    return __awaiter(this, void 0, void 0, function* () {
        yield teaSDK.collect(eventName, param, header);
    });
}
exports.sendEvent = sendEvent;//# sourceMappingURL=https://trae.private/sourcemaps/424b4bd987c6c6a4cadf1a08da420593cc6bf75a/node_modules/@ies/tron-client/dist/core/tea.js.map
